NOTE: As this notebook accesses protected TOPMed genetic data files, this notebook will only work if run in the gxpa-lcms-mediation workspace.
```{r}
library(data.table)
if(!file.exists("bcftools/bcftools")) system("git clone --recurse-submodules https://github.com/samtools/htslib.git; git clone https://github.com/samtools/bcftools.git; cd bcftools; make")

vs  <- fread(cmd='gcloud storage cat gs://fc-secure-4a392455-5587-4d6f-b8bd-01a1f834ae63/variants_of_interest.csv')
drs <- fread(cmd='gcloud storage cat gs://fc-secure-4a392455-5587-4d6f-b8bd-01a1f834ae63/mesa_geno_files_drs.csv')
dt  <- drs[vs,on='chr']
```

```{r}
get_dosage <- \(vcf_url, idx_fnm, chr) {
  fread(cmd = paste0(
    "bcftools/bcftools view",
      " -R ",chr,"_var_regions.txt",
      " -i'ID=@",chr,"_var_ids.txt'",
      " '",vcf_url,"##idx##",idx_fnm,"'", # vcf_url could have funny characters, so wrap in single quotes so the shell won't interpret them.
    " | ",
    "bcftools/bcftools +bcftools/plugins/dosage.so"
))}

# Extract genotypes (from both the HMB and HMB-NPU consent group files)
dosages <- dt[, by=chr, {
  writeLines(paste0(chr,'\t',pos), paste0(chr[1],'_var_regions.txt'))
  writeLines(rsid,                 paste0(chr[1],'_var_ids.txt'    ))

  # HMB                                                                     # HMB-NPU
  vcf_hmb_drs <- vcf_hmb_drs[1];                                            vcf_npu_drs <- vcf_npu_drs[1]
  csi_hmb_drs <- csi_hmb_drs[1];                                            csi_npu_drs <- csi_npu_drs[1]
  vcf_hmb_url <- system(paste('tnu drs access', vcf_hmb_drs), intern=T);    vcf_npu_url <- system(paste('tnu drs access', vcf_npu_drs), intern=T)
  csi_hmb_url <- system(paste('tnu drs access', csi_hmb_drs), intern=T);    csi_npu_url <- system(paste('tnu drs access', csi_npu_drs), intern=T)
  system(paste0("curl '",csi_hmb_url,"' -o idx_hmb.csi"));                  system(paste0("curl '",csi_npu_url,"' -o idx_npu.csi")) # Download VCF index files to allow fast random access.
  dosage_hmb <- get_dosage(vcf_hmb_url,  'idx_hmb.csi', chr[1]);            dosage_npu <- get_dosage(vcf_npu_url,  'idx_npu.csi', chr[1])

  dosage <- merge(dosage_hmb, dosage_npu, all=T)
}]

dosages <- vs[
  ][dosages, on=c(chr='#[1]CHROM', pos='[2]POS')
  ][, cpaid := paste(chr, pos, `[3]REF`, `[4]ALT`, sep='_') # Explicit chr_pos_ref_alt IDs, so will be robust to multiallelic sites.
  ][, `:=`(gene=NULL, rsid=NULL, chr=NULL, i.chr=NULL, pos=NULL, analysis=NULL, `[3]REF`=NULL, `[4]ALT`=NULL)
] |> transpose(make.names='cpaid', keep.names='NWD_Id')
dosages[, NWD_Id := sub('^\\[[0-9]+\\]', '', NWD_Id) ] # "[123]NWD987" -> "NWD987"

fwrite(dosages, 'dosages.csv')
system('gcloud storage cp dosages.csv $WORKSPACE_BUCKET')
```
