```{css echo=F}
pre {
  overflow-x: scroll;
}
pre code {
  white-space: pre;
}
```

# Setup
```{r}
library(data.table)
library(circlize)
library(ComplexHeatmap)
library(ggplot2)
library(ggrepel)
library(parallel)

data <- fread("analysis_df.csv")
```


# Heatmap of mPCs vs. covars
```{r}
covars <- c(paste0("gPC", 1:11), "age", "bmi", 
            "smoking", "gender_f0m1",
            "race", "site", "season", "month")

runs <- CJ(covars, mPC=paste0("mPC",1:20), sorted=F)
runs[, lm_res    := mapply(mPC,covars, FUN= \(mPC,covar) lm(data[[mPC]] ~ data[[covar]]))
   ][, anova_res := lapply(lm_res, anova)
   ][, p         := sapply(anova_res, \(res) res$`Pr(>F)`[1])
]
p_mtx <- as.matrix(dcast(runs, covars~mPC, value.var="p"), rownames=1)
p_mtx <- p_mtx[,paste0("mPC",1:20)]

# TODO ask why the two heatmaps are done differently:
# Covars, as above: lm(mPC ~ one_covar)
# SNPs, glm(snp ~ all_mPCs, family="binomial")

Heatmap(-log10(p_mtx),
  cell_fun = \(j,i,x,y,w,h,col) if(p_mtx[i,j]<0.05) grid.text("*",x,y,gp=gpar(col="white")),
  col = colorRamp2(c(0,10), c("black","steelblue")),
  cluster_columns=F, cluster_rows=F
)
```

# MWAS on genotypes and exposures
```{r mwas, cache=T}
met_names <- names(fread("metabolomics/QCd/merged_QCd.csv",nrows=0,drop=1))
vars2test <- c("rs295849","pa_bin","pa","mod_pa","vig_pa","mvpa")
covar_sets <- list(
  c("site", "gender_f0m1", "age"                                                                                                        ),
  c("site", "gender_f0m1", "age", "ses_score", "income_cat"                                                                             ),
  c("site", "gender_f0m1", "age", "ses_score", "income_cat", "drinks_per_week", "smoking", "ahei_score", "dash_score"                   ),
  c("site", "gender_f0m1", "age", "ses_score", "income_cat", "drinks_per_week", "smoking", "ahei_score", "dash_score", "race"           ),
  c("site", "gender_f0m1", "age", "ses_score", "income_cat", "drinks_per_week", "smoking", "ahei_score", "dash_score", paste0("gPC", 1:5))
)

met_eigvals <- prcomp(data[,..met_names])$sdev^2
met_eff_n <- sum(met_eigvals)^2 / sum(met_eigvals^2)
p_thresh <- 0.05/met_eff_n

  # Define all possible model combinations of met ~ var + covars
mwas <- CJ(met_names, vars2test, covar_sets, sorted=F)[
  # Run linear models
    , lm_coefs := mcmapply(met_names,vars2test,covar_sets, FUN = \(met,var,covar_set)
        lm(data[[met]] ~ ., data[,c(var,covar_set), with=F]) |> (\(x) summary(x)$coefficients)(),
        SIMPLIFY=F,
        mc.cores=6)

  # Extract results we want
  ][, var_coefs := mapply(lm_coefs, vars2test, FUN = \(coefs,var)
        coefs[rownames(coefs)==var,],
        SIMPLIFY=F)

  ][, estimate := sapply(var_coefs, "[", "Estimate"  )
  ][, se       := sapply(var_coefs, "[", "Std. Error")
  ][, p        := sapply(var_coefs, "[", "Pr(>|t|)"  )
  ][, q        := p.adjust(p, method="BH")
  ][, signif   := p < p_thresh

  # Cleanup
  ][, var_coefs := NULL
  ][,  lm_coefs := NULL
  ][, covar_sets := sapply(covar_sets,paste,collapse='+')
]
```

# TMP
```{r}
#tmp <- mwas[vars2test=="rs295849" & met_names %in% c("QI3513_HIL-pos","QI984_HIL-pos","QI7609_HIL-pos")]
#tmp <- mwas[signif==T]
#
mwas_gpc <- mwas[grepl("gPC",covar_sets)]
knitr::kable(mwas_gpc[signif==T, !c("covar_sets")][order(vars2test,p)])
#tmp <- mwas[grepl("gPC",covar_sets) & signif==T]
#
## Checking vs. old results
#mwas_gpc[vars2test=="rs295849" & signif==T][order(p)]
#mwas_gpc[vars2test=="rs295849" & met_names %in% c("QI3513_HIL-pos","QI984_HIL-pos","QI7609_HIL-pos")]
#
##qqnorm(mwas_qtl_gpc[order(p),-log10(p)])
```

# QQ
```{r}
mwas_gpc[
    order(p)
  ][, nlp := -log10(p)
  ][, nlp_expect := -log10(ppoints(.N)), by=vars2test
] |>
  ggplot(aes(x=nlp_expect, y=nlp, color=vars2test)) +
    geom_point() +
    geom_abline(slope=1, intercept=0) +
    theme_bw()

# Genomic inflation λ
calc_λ  <- function(x, p=0.5){
  x <- x[!is.na(x)]
  x.quantile <- quantile(x, p)
  round(qchisq(1 - x.quantile, 1) / qchisq(p, 1), 2)
}
mwas_gpc[, calc_λ(p), by=vars2test]
```

# Volcano
```{r warning=F, fig.width=16, fig.height=12}
ggplot(mwas_gpc, aes(x=estimate, y=-log10(p), color=p<p_thresh)) +
  geom_hline(yintercept=-log10(p_thresh), linetype="dashed") +
  geom_point(alpha=0.4, shape=16) +
  geom_text_repel(data=mwas_gpc[p<p_thresh], aes(label=met_names), size=3, show.legend=F) +
  scale_color_manual(name="Significant?\n(p<0.05/n_eff_metabolites)",values=c("FALSE"="black","TRUE"="red")) +
  facet_wrap(vars(vars2test), scales="free") +
  guides(color = guide_legend(override.aes = list(size=5)))
```



# Interaction analysis
lm(hdl ~ met*(snp or exposure) + covars), using MWAS's significant metabolites.
```{r}
#TODO
```

# Mediation analysis
```{r}
#TODO
```


# TODO
# QTL & Exposure MWASes ✓
Remem to use mvpa variable 
## QQs ✓
## Volcanoes ✓
## cutoff at 0.05/n_eff_metabs ✓

# Interaction analysis (hdl ~ met*(snp OR e) + covars) on the signif mets from MWAS
# Mediation analysis 

Most of this is in the univariate-analysis script.

# (later) Sensitivity analysis (hdl ~ (e or snp) + covars)
