```{css echo=F}
pre {
  overflow-x: scroll;
}
pre code {
  white-space: pre;
}
```

# Setup
```{r setup, message=F}
library(data.table)
library(circlize)
library(ComplexHeatmap)
library(ggplot2)
library(ggrepel)
library(mediation)
library(parallel)

data <- fread('analysis_df.csv')
```


# Heatmap of mPCs vs. covars
Note that not all samples contributed to the mPCs — only samples with measurements for all metabolomics methods CP/CN/HP/AN could be used in PCA.
```{r hm, cache=2}
covars <- c(paste0('gPC', 1:11),
            'site', 'season', 'month',
            'age', 'bmi', 'gender_f0m1', 'race',
            'ses_score', 'ahei_score', 'dash_score',
            'income_cat', 'drinks_per_week', 'smoking')

runs <- CJ(covars, mPC=paste0('mPC',1:20), sorted=F)[
  ][, lm_res    := mapply(mPC,covars, FUN = \(mPC,covar) lm(data[[mPC]] ~ data[[covar]]))
  ][, anova_res := lapply(lm_res, anova)
  ][, p         := sapply(anova_res, \(res) res$`Pr(>F)`[1])
]
p_mtx <- as.matrix(dcast(runs, covars~mPC, value.var='p'), rownames=1)
p_mtx <- p_mtx[,paste0('mPC',1:20)]

# TODO also do other way of doing heatmap, see journal day 241112.
# Covars, as above: lm(mPC ~ one_covar)
# SNPs, glm(snp ~ all_mPCs, family='binomial')

Heatmap(-log10(p_mtx),
  cell_fun = \(j,i,x,y,w,h,col) if(p_mtx[i,j]<0.05) grid.text('*',x,y,gp=gpar(col='white')),
  col = colorRamp2(c(0,10), c('black','steelblue')),
  cluster_columns=F, cluster_rows=F,
  heatmap_legend_param = list(title = 'log10(p)\nfrom lm')
)
```

# MWAS on genotypes and exposures
Model: `lm(met ~ [snp or exposure] + covars)`
```{r mwas, cache=T}
met_names <- names(fread('metabolomics/QCd/merged_QCd.csv',nrows=0,drop=1))
vars2test <- c('rs295849','pa_bin','pa','mod_pa','mvpa','vig_pa')
covar_sets <- list(
  c('site', 'gender_f0m1', 'age'                                                                                                        ),
  #c('site', 'gender_f0m1', 'age', 'ses_score', 'income_cat'                                                                             ),
  #c('site', 'gender_f0m1', 'age', 'ses_score', 'income_cat', 'drinks_per_week', 'smoking', 'ahei_score', 'dash_score'                   ),
  #c('site', 'gender_f0m1', 'age', 'ses_score', 'income_cat', 'drinks_per_week', 'smoking', 'ahei_score', 'dash_score', 'race'           ),
  c('site', 'gender_f0m1', 'age', 'ses_score', 'income_cat', 'drinks_per_week', 'smoking', 'ahei_score', 'dash_score', paste0('gPC', 1:5))
)

p_thresh <- local({
  met_mtx <- data[ rowSums(is.na(data[,..met_names]))==0, ..met_names ] |> as.matrix(rownames=1) # Only samples with data for ALL metabolomics methods
  met_eigvals <- prcomp(met_mtx, scale=T, center=T)$sdev^2
  met_eff_n <- sum(met_eigvals)^2 / sum(met_eigvals^2)

  message('# of effective / total metabolites: ',round(met_eff_n),'/',ncol(met_mtx))
  0.05/met_eff_n
})

  # Define all possible model combinations of met ~ var + covars
mwas <- CJ(met_names, vars2test, covar_sets, sorted=F)[
  # Run linear models
  ][, lm_coefs := mcmapply(met_names,vars2test,covar_sets, FUN = \(met,var,covar_set)
        lm(data[[met]] ~ ., data[,c(var,covar_set), with=F]) |> (\(model) summary(model)$coefficients)(),
      SIMPLIFY=F, mc.cores=4)

  # Extract results we want
  ][, var_coefs := mapply(lm_coefs, vars2test, FUN = \(coefs,var)
        coefs[rownames(coefs)==var,],
      SIMPLIFY=F)

  ][, estimate := sapply(var_coefs, '[', 'Estimate'  )
  ][, se       := sapply(var_coefs, '[', 'Std. Error')
  ][, p        := sapply(var_coefs, '[', 'Pr(>|t|)'  )
  ][, q        := p.adjust(p, method='BH')
  ][, signif   := p < p_thresh

  # Cleanup
  #][, covar_sets := sapply(covar_sets,paste,collapse='+')
  ][, var_coefs := NULL
  ][,  lm_coefs := NULL
]
```
```{r}
# Continuing with results from the run with gPC covars
mwas_gpc <- mwas[grepl("gPC",covar_sets)]
knitr::kable( mwas_gpc[signif==T][order(vars2test,p)] )
```

# QQ
```{r qq, cache=2}
mwas_gpc[
  ][ order(p)
  ][, nlp := -log10(p)
  ][, nlp_expect := -log10(ppoints(.N)), by=vars2test
] |>
  ggplot(aes(x=nlp_expect, y=nlp, color=vars2test)) +
    geom_point() +
    geom_abline(slope=1, intercept=0) +
    theme_bw()

# Genomic inflation λ
calc_λ  <- function(x, p=0.5){
  x <- x[!is.na(x)]
  x.quantile <- quantile(x, p)
  round(qchisq(1 - x.quantile, 1) / qchisq(p, 1), 2)
}
mwas_gpc[, calc_λ(p), by=vars2test]
```

# Volcano
```{r volc, warning=F, fig.width=16, fig.height=12, cache=2}
ggplot(mwas_gpc, aes(x=estimate, y=-log10(p), color=p<p_thresh)) +
  geom_hline(yintercept=-log10(p_thresh), linetype="dashed") +
  geom_point(alpha=0.4, shape=16) +
  geom_text_repel(data=mwas_gpc[p<p_thresh], aes(label=met_names), size=3, show.legend=F) +
  scale_color_manual(name="Significant?\n(p<0.05/n_eff_metabolites)",values=c("FALSE"="black","TRUE"="red")) +
  facet_wrap(vars(vars2test), scales="free") +
  guides(color = guide_legend(override.aes = list(size=5)))
```



# Mediation Analysis
Comparing models `lm(met ~ [snp or exposure] + covars)` and `lm(hdl_log ~ met + [snp or exposure] + covars)` to find mediators.
```{r mediation, cache=T}
mediation_results <-
  mwas_gpc[
    ][ signif==T # ↓ No mcmapply, mediate() is already multithreaded 
    ][, med_res := mapply(met_names,vars2test,covar_sets,.I, FUN = \(met,var,covar_set,i) {
          #cat(i,'/',.N,'\r') # Progress bar
          data <- data[!is.na(hdl_log)]
          mediate(
            lm(data[[   met   ]] ~ ., data[,c(    var,covar_set),with=F]),
            lm(data[['hdl_log']] ~ ., data[,c(met,var,covar_set),with=F]),
            treat=var, mediator=met, robustSE=T, sims=1000
          )[c('d0.p', 'n0', 'nobs')] # Select only the results we need. Otherwise eats up a lot of memory.
        }, SIMPLIFY=F)

    # Extract results we want
    ][, mwas_p    := p
    ][, acme_p    := sapply(med_res,'[[','d0.p')
    ][, med_prop  := sapply(med_res,'[[','n0'  )
    ][, n_samples := sapply(med_res,'[[','nobs')

    # Cleanup
    #][, covar_sets := sapply(covar_sets,paste,collapse='+')
    ][, `:=`(covar_sets=NULL, med_res=NULL, signif=NULL, estimate=NULL, se=NULL, p=NULL, q=NULL)
    ]
```
```{r mediation_tbl}
knitr::kable( mediation_results[acme_p<0.05][order(vars2test,acme_p,mwas_p)] )
```



# Interaction analysis
Model: `lm(hdl ~ met*[snp or exposure] + covars)`, using MWAS's significant metabolites.
```{r intrx, cache=T}
interaction_results <-
  mwas_gpc[
    ][ signif==T
    ][, lm_coefs := mcmapply(met_names,vars2test,covar_sets,.I, FUN = \(met,var,covar_set,i) {
          #cat(i,'/',.N,'\r') # Progress bar
          data <- data[!is.na(hdl_log)]
          lm(data[['hdl_log']] ~ data[[met]] * data[[var]] + ., data[,covar_set,with=F]) |> (\(model) summary(model)$coefficients)()
        }, SIMPLIFY=F, mc.cores=4)

    # Extract results we want
    ][, intrx_coefs := lapply(lm_coefs, \(coefs)
          coefs[rownames(coefs)=='data[[met]]:data[[var]]',])

    ][, estimate := sapply(intrx_coefs, '[', 'Estimate'  )
    ][, se       := sapply(intrx_coefs, '[', 'Std. Error')
    ][, p        := sapply(intrx_coefs, '[', 'Pr(>|t|)'  )
    ][, q        := p.adjust(p, method='BH')

    # Cleanup
    #][, covar_sets := sapply(covar_sets,paste,collapse='+')
    ][, `:=`(covar_sets=NULL, lm_coefs=NULL, intrx_coefs=NULL, signif=NULL)
    ]
```
```{r intrx_tbl}
knitr::kable( interaction_results[p<0.05][order(vars2test,p)] )
```


# (later) Sensitivity analysis (hdl ~ (e or snp) + covars)


# Moderated Mediation
```{r moderated, include=F}
# TODO just copied the old code
#test_moderated_mediation <- function(x, m, y, w, covars) {
#    # For now, w (the moderating variable) must be binary (0/1)
#    med_form_str <- paste0("m ~ x * w + ", paste(covars, collapse=" + "))
#    med_fit <- lm(as.formula(med_form_str), data=analysis_df_lcms)
#    out_form_str <- paste0("y ~ m * w + x * w + ", paste(covars, collapse=" + "))
#    out_fit <- lm(as.formula(out_form_str), data=analysis_df_lcms)
#    med_out <- mediate(med_fit, out_fit, 
#                       treat="x", mediator="m",
#                       robustSE=TRUE, sims=10)  # N_sims doesn't need to be high for this step
#    modmed_out <- test.modmed(med_out, list(w=0), list(w=1), sims=100)
#    modmed_out
#}
```
