```{css echo=F}
pre {
  overflow-x: scroll;
}
pre code {
  white-space: pre;
}
```
# Setup
```{r setup, message=F}
library(data.table)
library(ggplot2)
library(ggrepel)
library(lme4)
library(lmerTest) # so lmer() returns p-vals
library(mediation)
library(parallel)

options(mc.cores=6L)

options(datatable.na.strings=c('NA',''))
data <- fread('analysis_df.csv')
variants <- grep('^rs[0-9]+', names(data), value=T)
data[
  ][, mesa_id      := as.factor(mesa_id)
  ][, exam         := as.factor(exam)
  ][, names(.SD)   := lapply(.SD, as.character), .SDcols=variants # Variants are numeric (0/1/2), but we temporarily set to character so they don't get scaled.
  ][, names(.SD)   := lapply(.SD, scale       ), .SDcols=is.numeric
  ][, names(.SD)   := lapply(.SD, as.numeric  ), .SDcols=variants
  ][, paste0(names(.SD),'_bin') := lapply(.SD, \(geno) fifelse(geno==0,0,1)), .SDcols=variants # test.modmed only supports binary covariates apparently, so make binary versions of the variants.
]
data_exam1 <- data[exam==1]
data_exam2 <- data[exam==2]
data_exam3 <- data[exam==3]
data_exam4 <- data[exam==4]
data_exam5 <- data[exam==5]

numeric_colnms  <- names(data)[sapply(data,is.numeric)]
baseline_colnms <- paste0(numeric_colnms,'_0')
delta_colnms    <- paste0(numeric_colnms,'_Δ')
data[, (baseline_colnms) := .SD[exam==1],    .SDcols=numeric_colnms,     by='mesa_id'              ]
data[, (   delta_colnms) := data[,.SD,.SDcols=numeric_colnms] - data[,.SD,.SDcols=baseline_colnms] ]

#data[mesa_id %in% mesa_id[duplicated(mesa_id)], .SD, .SDcols=grepl('QI19074_HI|exam',names(data)), by='mesa_id'] # Sanity check

met_nms <- names(fread('metabolomics/QCd/merged_QCd.csv',nrows=0,drop=1))
covars <- c('site', 'sex', 'age', 'ses_score', 'income', 'drinks_per_week', 'smoking', 'ahei_score', 'dash_score', paste0('gPC', 1:5), 'race')

# Data table that we will gradually fill with the results of each analysis
mwas <- fread('metabolomics/met_info.csv')[unique_met_id %in% met_nms, .(Method,MZ,RT,HMDB_ID,Metabolite,unique_met_id)]

calc_eff_n_metabolites <- \(met_nms) {
  met_mtx <- data[ rowSums(is.na(data[,..met_nms]))==0, ..met_nms ] |> as.matrix(rownames=1) # Only samples with data for ALL metabolomics methods
  met_eigvals <- prcomp(met_mtx, scale=T, center=T)$sdev^2
  met_eff_n <- sum(met_eigvals)^2 / sum(met_eigvals^2)
}
```

# MWAS on genotypes and exposures
Models:\
* __Linear Model__  only  using exam 1   data: `lm  (  met   ~  mvpa_wins   + covars)`\
* __Linear Mixed Model__  using exam 1-5 data: `lmer(  met   ~  mvpa_wins   + covars + (1 | mesa_id))`\
* __Interaction Model__   using exam 1   data: `lm  (hdl_log ~ met*rs295849 + covars + sex*rs295849 )`\
* __Mediation Model__     using exam 1-5 data: Comparing models:\
  `lm(  met   ~       mvpa_wins + covars)` and\
  `lm(hdl_log ~ met + mvpa_wins + covars)`\
  to find metabolites which mediate the hdl ~ pa relationship.
* __Moderated Mediation__ using exam 1-5 data: Comparing models:\
  `lm(  met       ~                mvpa_wins*rs295849_bin + covars)` and\
  `lm(hdl_log ~ met*rs295849_bin + mvpa_wins*rs295849_bin + covars)`.
TODO explain the other misc models
```{r mwas, cache=T, cache.depend=digest::digest(mwas), results='hide', eval=F}
mwas <- mwas[
  # Define models for the main analysis.
  ## LM
  ][,                lm_fmla               :=  paste0(unique_met_id,' ~ ',                                        'mvpa_wins',                 ' + ',paste(covars,collapse='+'))
  ][,            med_lm_fmla_m             :=  paste0(unique_met_id,' ~ ',                                        'mvpa_wins',                 ' + ',paste(covars,collapse='+'))
  ][,            med_lm_fmla_y             :=  paste0(  'hdl_log'  ,' ~ ',unique_met_id,                    ' + ','mvpa_wins',                 ' + ',paste(covars,collapse='+'))
  ][, paste0( 'intrx_lm_fmla_',  variants) :=  lapply(variants, \(variant) paste0(  'hdl_log',  ' ~ ',unique_met_id,'*',variant,       ' + ',      'sex*',variant,       ' + ',paste(covars,collapse='+')))
  ][, paste0('modmed_lm_fmla_m_',variants) :=  lapply(variants, \(variant) paste0(unique_met_id,' ~ ',                                       'mvpa_wins*',variant,'_bin',' + ',paste(covars,collapse='+')))
  ][, paste0('modmed_lm_fmla_y_',variants) :=  lapply(variants, \(variant) paste0(  'hdl_log',  ' ~ ',unique_met_id,'*',variant,'_bin',' + ','mvpa_wins*',variant,'_bin',' + ',paste(covars,collapse='+')))

  ## LMM (basically just adding a random intercept term to the LM formulas.)
  ][,                lmm_fmla               := paste(lm_fmla,       ' + (1|mesa_id)')
  ][,            med_lmm_fmla_m             := paste(med_lm_fmla_m, ' + (1|mesa_id)')
  ][,            med_lmm_fmla_y             := paste(med_lm_fmla_y, ' + (1|mesa_id)')
  ][, paste0( 'intrx_lmm_fmla_',  variants) := lapply(.SD, paste,   ' + (1|mesa_id)'), .SDcols=paste0( 'intrx_lm_fmla_',  variants)
  ][, paste0('modmed_lmm_fmla_m_',variants) := lapply(.SD, paste,   ' + (1|mesa_id)'), .SDcols=paste0('modmed_lm_fmla_m_',variants) # Unfortunately test.modmed can't handle LMMs yet: https://github.com/kosukeimai/mediation/issues/69
  ][, paste0('modmed_lmm_fmla_y_',variants) := lapply(.SD, paste,   ' + (1|mesa_id)'), .SDcols=paste0('modmed_lm_fmla_y_',variants) # ^

  # Other models of interest.
  ## The GxM LMM above, except with the metabolite term decomponsed into...  ...baseline ↓ (exam1)...          ...and delta ↓ (examX - exam1) terms. The baseline term should be concordant with the LM form. The delta term should show us the effect of the additional longitudinal data.
  ][, paste0('Δ_intrx_lmm_fmla_',variants) := lapply(variants, \(variant) paste0(  'hdl_log',  ' ~ ',unique_met_id,'_0*',variant,' + ',unique_met_id,'_Δ*',variant,' + ','sex*',variant,' + ',paste(covars,collapse='+'),' + (1|mesa_id)'))
  ## Metabolite ~ genotype main effect.
  ][, paste0( 'g_main_lmm_fmla_',variants) := lapply(variants, \(variant) paste0(unique_met_id,' ~ ',variant,' + ',paste(covars,collapse='+'),' + (1|mesa_id)'))
  ## Checking for metabolites' nonlinear effects on HDL, which could confound mediation.
  ][, sq_lmm_fmla := paste0('hdl_log ~ ',unique_met_id,' + I(',unique_met_id,'^2) + ',paste(covars,collapse='+'),' + (1|mesa_id)')
]

# One-off formulas (not run for every metabolite).
## TODO not super readable... not sure how to make it so while still automated. Only other option I can think of is assign() but that's probably uglier.
one_offs <- data.table()[
  ## Sanity check confirming the obvious HDL~MVPA association exists.
  ][, hdl_pa_lm_fmla  := paste0('hdl_log ~ mvpa_wins + ', paste(covars,collapse='+'))
  ][, hdl_pa_lmm_fmla := paste(hdl_pa_lm_fmla,'+ (1|mesa_id)')
  ## Confirming we can reproduce the HDL~SNP*MVPA interaction this whole analysis is based on.
  ][, paste0(      'hdl_gxpa_lm_fmla_', variants) := as.list(paste0('hdl_log ~ ',variants,'*mvpa_wins + ',variants,'*sex + ',paste(covars,collapse='+')                 ))
  ][, paste0('hdl_gxpa_nosex_lm_fmla_', variants) := as.list(paste0('hdl_log ~ ',variants,'*mvpa_wins + ',                   paste(covars,collapse='+')                 ))
  ][, paste0(      'hdl_gxpa_lmm_fmla_',variants) := as.list(paste0('hdl_log ~ ',variants,'*mvpa_wins + ',variants,'*sex + ',paste(covars,collapse='+'),' + (1|mesa_id)'))
  ][, paste0('hdl_gxpa_nosex_lmm_fmla_',variants) := as.list(paste0('hdl_log ~ ',variants,'*mvpa_wins + ',                   paste(covars,collapse='+'),' + (1|mesa_id)'))
] |> as.list(all.names=T)
```

# Run one-off models
```{r}
lapply(one_offs, \(fmla)
  tryCatch( lm  (fmla,data), error=\(e) # Kinda goofy code
            lmer(fmla,data)
  ) |> (\(model) summary(model)$coefficients)()
) |> suppressWarnings()
```

# Run metabolite models
```{r}
mwas <- mwas[
  # LM on all metabolites, exam 1 only.
  ][, c('lm_est', 'lm_se', 'lm_t', 'lm_p') := transpose(mcMap(lm_fmla, f=\(fmla) {
        summary( lm(fmla,data_exam1) )$coefficients['mvpa_wins',]
      }))

  # LMM on all metabolites.
  ][, c('lmm_est', 'lmm_se', 'lmm_df', 'lmm_t', 'lmm_p') := transpose(mcMap(lmm_fmla, f=\(fmla) {
        summary( lmerTest::lmer(fmla,data) )$coefficients['mvpa_wins',] # lmerTest::lmer adds p-values
      }))

  ][, p_thresh_mwas    := 0.05/calc_eff_n_metabolites(unique_met_id)
  ][, p_thresh_med_lm  := 0.05/calc_eff_n_metabolites(unique_met_id[lm_p  < p_thresh_mwas])
  ][, p_thresh_med_lmm := 0.05/calc_eff_n_metabolites(unique_met_id[lmm_p < p_thresh_mwas])

  # LM mediation on metabolites significant from the LM MWAS, exam 1 only.
  ][lm_p < p_thresh_mwas
    , c('med_lm_acme_p','med_lm_prop','med_lm_n_samples') := transpose(mcMap(med_lm_fmla_m,med_lm_fmla_y,unique_met_id, f=\(fmla_m,fmla_y,met) {
        mediate(
          lm(fmla_m, data_exam1[!is.na(hdl_log)]),
          lm(fmla_y, data_exam1                 ),
          treat='mvpa_wins',
          mediator=met,
          robustSE=T,
          sims=500 # Enough sims to reach the p-val threshold
        )[c('d0.p', 'n0', 'nobs')]
      }))

  # LMM mediation on metabolites significant from the LMM MWAS.
  ][lmm_p < p_thresh_mwas
    , c('med_lmm_acme_p','med_lmm_prop','med_lmm_n_samples') := transpose(mcMap(med_lmm_fmla_m,med_lmm_fmla_y,unique_met_id, f=\(fmla_m,fmla_y,met) {
        mediate(
          lme4::lmer(fmla_m, data[!is.na(hdl_log)]), # Need to use lme4::lmer() not lmerTest::lmer(). lmerTest changes the output format s.t. mediate() doesn't recognize it anymore.
          lme4::lmer(fmla_y, data                 ),
          treat='mvpa_wins',
          mediator=met,
          robustSE=T,
          sims=500 # Enough sims to reach the p-val threshold
        )[c('d0.p', 'n0', 'nobs')]
      }))
  ][, .SDcols = \(col) any(lengths(col)==0), # NULL -> NA for rows with calculations skipped.
      (names(.SD)) := lapply(.SD, \(col) fifelse(lengths(col)==0, list(NA), col) |> unlist())

  # Other models
  ## LMM hdl_log ~ met + met^2 + covars (adding a nonlinear term)
  ][, c('sq_lmm_est', 'sq_lmm_se', 'sq_lmm_df', 'sq_lmm_t', 'sq_lmm_p') := transpose(mcMap(sq_lmm_fmla,paste0('I(',unique_met_id,'^2)'), f=\(fmla,term) {
        summary( lmerTest::lmer(fmla,data) )$coefficients[term,]
      }))
]
```

# Run genetic interaction models
```{r}
for(variant in variants) {
  intrx_term_pats   <- paste0(variant,':',mwas$unique_met_id,  '|',mwas$unique_met_id,  ':',variant)
  intrx_term_pats_0 <- paste0(variant,':',mwas$unique_met_id,'_0|',mwas$unique_met_id,'_0:',variant)
  intrx_term_pats_Δ <- paste0(variant,':',mwas$unique_met_id,'_Δ|',mwas$unique_met_id,'_Δ:',variant)

  mwas <- mwas[
    # LM GxM on all metabolites, exam 1 only.
    ][, .SDcols = paste0('intrx_lm_fmla_',variant)
      , paste0(c('intrx_lm_est_', 'intrx_lm_se_', 'intrx_lm_t_', 'intrx_lm_p_'), variant) :=
          transpose(mcMap(.SD[[1]],intrx_term_pats, f=\(fmla,pat) {
            coefs <- summary( lm(fmla,data_exam1) )$coefficients
            coefs[grepl(pat,rownames(coefs)),]
          }))

    # LMM GxM on all metabolites.
    ][, .SDcols = paste('intrx_lmm_fmla_',variant)
      , paste0(c('intrx_lmm_est_', 'intrx_lmm_se_', 'intrx_lmm_df_', 'intrx_lmm_t_', 'intrx_lmm_p_'), variant) :=
          transpose(mcMap(intrx_lmm_fmla,intrx_term_pats, f=\(fmla,pat) {
            coefs <- summary( lmerTest::lmer(fmla,data_exam) )$coefficients
            coefs[grepl(pat,rownames(coefs)),]
          }))

    ][, paste0('p_thresh_intrx_lm_', variant) := 0.05/calc_eff_n_metabolites(na.omit(unique_met_id[med_lm_acme_p  < p_thresh_med_lm ]))
    ][, paste0('p_thresh_intrx_lmm_',variant) := 0.05/calc_eff_n_metabolites(na.omit(unique_met_id[med_lmm_acme_p < p_thresh_med_lmm]))
    #][, p_thresh_modmed_lm  := 0.05/calc_eff_n_metabolites(na.omit(unique_met_id[intrx_lm_p < p_thresh_intrx_lm ]))
    ][, paste0('p_thresh_modmed_lm', variant) := 0.05 # Hardcoded 0.05 b/c there are very few mets w/ that make it this far, and that causes an error in SVD calculation.
    ][, paste0('p_thresh_modmed_lmm',variant) := 0.05/calc_eff_n_metabolites(na.omit(unique_met_id[intrx_lmm_p < p_thresh_intrx_lmm]))

    # Run LM moderated mediation on metabolites significant from the interaction LM, exam 1 only.
    ][intrx_lm_p  < p_thresh_intrx_lm  | # Actualaly run using both LM & LMM significant metabolites because test.modmed does not implement support for LMMs yet, so LM is the best we can do: https://github.com/kosukeimai/mediation/issues/69
      intrx_lmm_p < p_thresh_intrx_lmm 
      , paste0(c('modmed_lm_acme_p_','modmed_lm_ade_p_'),variant) :=
          transpose(mcMap(modmed_lm_fmla_m,modmed_lm_fmla_y,unique_met_id, f=\(fmla_m,fmla_y,met) {
            mediate(
              lm(fmla_m, data_exam1[!is.na(hdl_log)]),
              lm(fmla_y, data_exam1                 ),
              treat='mvpa_wins',
              mediator=met,
              robustSE=TRUE,
              sims=10 # N_sims doesn't need to be high for this step
            ) |>
            test.modmed(setNames(list(0),paste0(variant,'_bin')), # TODO ugly
                        setNames(list(1),paste0(variant,'_bin')), sims=100) |>
            (\(results) c(results[[1]]$p.value, results[[2]]$p.value))() # Get just the ACME & ADE p-vals
          }))
    ][, .SDcols = \(col) any(lengths(col)==0), # Again, NULL -> NA for rows with calculations skipped.
        (names(.SD)) := lapply(.SD, \(col) fifelse(lengths(col)==0, list(NA), col) |> unlist())


    # Other models
    ## LM GxM for other exams
    #][, .SDcols = paste0('intrx_lm_fmla_',variant)
    #  , paste0(c('intrx_lm_est_', 'intrx_lm_se_', 'intrx_lm_t_', 'intrx_lm_p_'), variant,'_2') :=
    #      transpose(mcMap(.SD[[1]],intrx_term_pats, f=\(fmla,pat) {
    #        coefs <- summary( lm(fmla,data_exam2) )$coefficients
    #        coefs[grepl(pat,rownames(coefs)),]
    #      }))

    ## LMM GxM with metabolite decomposed into baseline and delta terms
    #][# unique_met_id %in% c('QI48826_C18_neg','QI7292_C8_pos','QI46325_C18_neg','QI42688_C18_neg','QI54634_C18_neg','QI45506_C18_neg','QI5007_C8_pos','QI14471_HILIC_pos','QI3365_C8_pos','QI5103_C8_pos','QI13024_C18_neg')
    #  , paste0(c('intrx_lmm_est_0', 'intrx_lmm_se_0', 'intrx_lmm_df_0', 'intrx_lmm_t_0', 'intrx_lmm_p_0',
    #             'intrx_lmm_est_Δ', 'intrx_lmm_se_Δ', 'intrx_lmm_df_Δ', 'intrx_lmm_t_Δ', 'intrx_lmm_p_Δ'), variant) :=
    #      transpose(Map(Δ_intrx_lmm_fmla,intrx_term_pats_0,intrx_term_pats_Δ, f=\(fmla,pat0,patΔ) {
    #        coefs <- summary( lmerTest::lmer(fmla,data) )$coefficients
    #        unlist(c(coefs[grepl(pat0,rownames(coefs)),],
    #                 coefs[grepl(patΔ,rownames(coefs)),]))
    #      }))

    # Run LMM genetic main effect model (met ~ SNP + covars)
    #][, paste0(c('main_lmm_est', 'main_lmm_se', 'main_lmm_df', 'main_lmm_t', 'main_lmm_p'), variant) :=
    #      transpose(mcMap(g_main_lmm_fmla, f=\(fmla) {
    #        summary( lmerTest::lmer(fmla,data) )$coefficients[variant,]
    #      }))
  ]
}
```






# Junk
```{r}
#load('250112_modmed_lmm_done.RData')
#load('250224-modemed_lmm_done-delta_cols.RData')
#load('250227-mwas_mainG-mwas_sq.RData')
load('250320.RData')

{ mwas[      lm_p < .05, plot(      lm_est/      lm_se,       lmm_est/      lmm_se, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM MWAS Z-scores\n       (metabolites with MWAS LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }
{ mwas[intrx_lm_p < .05, plot(intrx_lm_est/intrx_lm_se, intrx_lmm_est/intrx_lmm_se, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM Interaction Z-scores\n(metabolites with GxM  LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }
#mwas[lm_p < .1 & lm_est>0, boxplot(lm_est/lm_se, lmm_est/lmm_se, names=c('LM','LMM'), ylab='Z-score', main='Metabolites whose LM p-value was < 0.1\nPOSITIVE Z-scores')]
#mwas[lm_p < .1 & lm_est<0, boxplot(lm_est/lm_se, lmm_est/lmm_se, names=c('LM','LMM'), ylab='Z-score', main='Metabolites whose LM p-value was < 0.1\nNEGATIVE Z-scores')]
{ mwas[lm_p < .05, boxplot(abs(      lm_est/      lm_se), abs(      lmm_est/      lmm_se), names=c('LM','LMM'), ylab='|Z-score|', main='Metabolites whose LM p-value was < 0.1')] }
{ mwas[lm_p < .05, boxplot(abs(intrx_lm_est/intrx_lm_se), abs(intrx_lmm_est/intrx_lmm_se), names=c('LM','LMM'), ylab='|Z-score|', main='Metabolites whose **Interaction** LM p-value was < 0.1')] }
#mwas[                                  , plot(-log10(med_lm_acme_p+1e-24), -log10(med_lmm_acme_p+1e-24), col=rgb(0,0,0,0.2),pch=16)] + abline(0,1,col='red')
#mwas[                                  , plot(-log10(fifelse(is.na(med_lm_acme_p), 1,med_lm_acme_p +1e-24)),
#                                              -log10(fifelse(is.na(med_lmm_acme_p),1,med_lmm_acme_p+1e-24)),
#                                              col=rgb(0,0,0,0.2),pch=16)] + abline(0,1,col='red')

#cat(mwas[order(      lm_p)[1:30],       lm_p/      lmm_p],sep='\n')
#cat(mwas[order(intrx_lm_p)[1:30], intrx_lm_p/intrx_lmm_p],sep='\n')

{ mwas[      lm_p<.05, hist(log(abs(      lmm_est/      lmm_se)/abs(      lm_est/      lm_se)), xlab='log(LMM_Z / LM_Z)', breaks=40L, main='Metabolites whose LM p-value was < 0.1\nLog LMM/LM Z-score ratios') ]; abline(0,1e10,col='red') }
{ mwas[intrx_lm_p<.05, hist(log(abs(intrx_lmm_est/intrx_lmm_se)/abs(intrx_lm_est/intrx_lm_se)), xlab='log(LMM_Z / LM_Z)', breaks=40L, main='Metabolites whose **Interaction** LM p-value was < 0.1\nLog Interaction LMM/LM Z-score ratios') ]; abline(0,1e24,col='red')}

# Note that in later steps, the increased number of significant mets may just be because more were tested because they passed the threshold of the previous step.
# Although the p thresholds do control for the # of effective metabolites.
mwas[                                                                                            , sum(       lm_p       < p_thresh_mwas     )]
mwas[lm_p  < p_thresh_mwas                                                                       , sum(   med_lm_acme_p  < p_thresh_med_lm   )]
mwas[lm_p  < p_thresh_mwas & med_lm_acme_p  < p_thresh_med_lm                                    , sum( intrx_lm_p       < p_thresh_intrx_lm )]
mwas[lm_p  < p_thresh_mwas & med_lm_acme_p  < p_thresh_med_lm  & intrx_lm_p  < p_thresh_intrx_lm , sum(modmed_lm_acme_p  < p_thresh_modmed_lm)]
mwas[                                                                                            , sum(       lmm_p      < p_thresh_mwas     )]
mwas[lmm_p < p_thresh_mwas                                                                       , sum(   med_lmm_acme_p < p_thresh_med_lmm  )]
mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm                                   , sum( intrx_lmm_p      < p_thresh_intrx_lmm)]
mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm & intrx_lmm_p < p_thresh_intrx_lmm, sum(modmed_lm_acme_p  < p_thresh_modmed_lm)]
mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm & intrx_lmm_p < p_thresh_intrx_lmm, sum(modmed_lm_acme_p  < .1                )]

mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm & intrx_lmm_p < p_thresh_intrx_lmm, .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')]
mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm                                   , .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')][order()]
mwas[lmm_p < p_thresh_mwas                                                                       , .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')]

# If we pretend we meant to do an interaction mwas
mwas[                                                                                            , sum( intrx_lm_p       < p_thresh_mwas     )]
mwas[                                                                                            , sum( intrx_lmm_p      < p_thresh_mwas     )]
mwas[intrx_lm_p  < p_thresh_mwas                                                                 , .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')]
mwas[intrx_lmm_p < p_thresh_mwas                                                                 , .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')][order(intrx_lm_p)]

#tmp <- mwas[intrx_lmm_p < p_thresh_mwas                                                                 , .SD, .SDcols=patterns('met|_est|_p')][order(intrx_lm_p)]
"%ni%" <- Negate("%in%")
tmp <- mwas[order(unique_met_id %ni% c('QI48826_C18_neg','QI7292_C8_pos','QI46325_C18_neg','QI42688_C18_neg','QI54634_C18_neg','QI45506_C18_neg','QI5007_C8_pos','QI14471_HILIC_pos',  'QI3365_C8_pos','QI5103_C8_pos','QI13024_C18_neg'))]
```
```{r}
# TODO: knitr tables like this: knitr::kable(mwas[lm_p   <p_thresh_mwas, .(MZ,RT,HMDB_ID,Metabolite,unique_met_id,lm_p   )][order(lm_p   )] )
```

## QQ
```{r qq}
mwas[
  ][ order(lmm_p)
  ][, nlp := -log10(lmm_p)
  ][, nlp_expect := -log10(ppoints(.N))
] |>
  ggplot(aes(x=nlp_expect, y=nlp)) +
    geom_point() +
    geom_abline(slope=1, intercept=0) +
    theme_bw()

# Genomic inflation λ
calc_λ  <- function(x, p=0.5){
  x <- x[!is.na(x)]
  x.quantile <- quantile(x, p)
  round(qchisq(1 - x.quantile, 1) / qchisq(p, 1), 2)
}
mwas[, calc_λ(lmm_p)]
```

## Volcano
```{r volc, warning=F, fig.width=12, fig.height=8}
ggplot(mwas, aes(x=lmm_est, y=-log10(lmm_p), color=lmm_p<p_thresh_mwas)) +
  geom_hline(yintercept=-log10(p_thresh_mwas), linetype="dashed") +
  geom_point(alpha=0.4, shape=16) +
  geom_text_repel(data=mwas[lmm_p==p_thresh_mwas], aes(label=Metabolite), size=3, show.legend=F) +
  scale_color_manual(name="Significant?\n(p<0.05/n_eff_metabolites)",values=c("FALSE"="black","TRUE"="red")) +
  guides(color = guide_legend(override.aes = list(size=5)))
```


```{r}
mwas[unique_met_id%in%c('QI3365_C8_pos', 'QI5103_C8_pos', 'QI13024_C18_neg', 'QI48826_C18_neg'),
     .(unique_met_id,MZ,RT,Metabolite,lm_p,lmm_p,intrx_lmm_p,intrx_lmm_p,med_lmm_prop)
]


cor(mwas$intrx_lmm_est,mwas$intrx_lmm_est_0,use='complete.obs')
cor(mwas$intrx_lmm_est,mwas$intrx_lmm_est_Δ,use='complete.obs')
cor(abs(mwas_new$intrx_lmm_est),abs(mwas_new$intrx_lmm_est_0),use='complete.obs')
cor(abs(mwas_new$intrx_lmm_est),abs(mwas_new$intrx_lmm_est_Δ),use='complete.obs')
ggplot(mwas_new, aes(x=intrx_lmm_est/intrx_lmm_se,y=intrx_lmm_est_0/intrx_lmm_se_0)) + geom_point() + geom_text_repel(data=mwas_new, aes(label=unique_met_id), size=3, show.legend=F) + geom_abline(slope=1)
ggplot(mwas_new, aes(x=intrx_lmm_est/intrx_lmm_se,y=intrx_lmm_est_Δ/intrx_lmm_se_Δ)) + geom_point() + geom_text_repel(data=mwas_new, aes(label=unique_met_id), size=3, show.legend=F) + geom_abline(slope=1)

ggplot(mwas_new, aes(x=intrx_lmm_est_0/intrx_lmm_se_0,y=intrx_lmm_est_Δ/intrx_lmm_se_Δ)) + geom_point() + geom_text_repel(data=mwas, aes(label=unique_met_id), size=3, show.legend=F) + geom_abline(slope=1)

plot(abs(mwas$intrx_lmm_est),abs(mwas$intrx_lmm_est_0), main='baseline')
plot(abs(mwas$intrx_lmm_est),abs(mwas$intrx_lmm_est_Δ), main='delta')
#unique_met_id %in% c('QI3365_C8_pos', 'QI5103_C8_pos', 'QI13024_C18_neg', 'QI48826_C18_neg', 'QI7292_C8_pos', 'QI46325_C18_neg', 'QI42688_C18_neg', 'QI54634_C18_neg', 'QI45506_C18_neg', 'QI5007_C8_pos', 'QI14471_HILIC_pos')



{ mwas[unique_met_id %in% mets, plot(      lm_est/      lm_se,       lmm_est/      lmm_se, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM MWAS Z-scores\n       (metabolites with MWAS LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }
{ mwas[unique_met_id %in% mets, plot(intrx_lm_est/intrx_lm_se, intrx_lmm_est/intrx_lmm_se, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM Interaction Z-scores\n(metabolites with GxM  LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }

{ mwas[unique_met_id %in% mets, plot(      lm_est,       lmm_est, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM MWAS Z-scores\n       (metabolites with MWAS LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }
{ mwas[unique_met_id %in% mets, plot(intrx_lm_est, intrx_lmm_est, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM Interaction Z-scores\n(metabolites with GxM  LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }


# df w/ ests and z for LM & LMM, decomposed lmm baseline and delta... 
# Plot w/ the 11 mets as the x axis, and above each is a colored dot for each of the above stats.
# will prolly ned long format
mets <- c('QI48826_C18_neg','QI7292_C8_pos','QI46325_C18_neg','QI42688_C18_neg','QI54634_C18_neg','QI45506_C18_neg','QI5007_C8_pos','QI14471_HILIC_pos',  'QI3365_C8_pos','QI5103_C8_pos','QI13024_C18_neg')
df <- mwas[
  ][ unique_met_id %in% mets
    , .(unique_met_id,
        intrx_lm_est,
        intrx_lmm_est_0,
        intrx_lmm_est,
        intrx_lmm_est_Δ,
        intrx_lm_z    = intrx_lm_est   /intrx_lm_se   ,
        intrx_lmm_z_0 = intrx_lmm_est_0/intrx_lmm_se_0,
        intrx_lmm_z   = intrx_lmm_est  /intrx_lmm_se  ,
        intrx_lmm_z_Δ = intrx_lmm_est_Δ/intrx_lmm_se_Δ
      )
] #|>
  melt(id.vars='unique_met_id')

ggplot(df[grepl('z',variable)], aes(x=unique_met_id,y=value,fill=variable)) + geom_bar(stat='identity',position='dodge') + theme(axis.text.x=element_text(angle=45, hjust = 1))
ggplot(df[grepl('est',variable)], aes(x=unique_met_id,y=value,fill=variable)) + geom_bar(stat='identity',position='dodge') + theme(axis.text.x=element_text(angle=45, hjust = 1))


ggplot(mwas_new, aes(x=unique_met_id,y=intrx_lm_est)) + geom_bar(stat='identity',position='dodge')

df[,unique_met_id:=NULL]
df[, names(.SD) := NULL, .SDcols=patterns('z')]
setnames(df,c('LM estimates','LMM estimates (baseline)','LMM estimates','LMM estimates (delta)'))#, 'LM z-scores','LMM z-scores (baseline)','LMM z-scores','LMM z-scores (delta)'))
Heatmap(
  cor_mtx <- cor(df,method='pearson'),
  col=colorRamp2(c(0.5, 1), c("white", "red")),
  cell_fun = \(j,i,x,y,w,h,col) grid.text(cor_mtx[i,j]|>round(digits=2),x,y,gp=gpar(fontsize=10)),
  show_heatmap_legend=F,
  column_title='Pearson correlations'
)




mwas_old <- copy(mwas)

mwas_new <- mwas[unique_met_id %in%  c('QI48826_C18_neg','QI7292_C8_pos','QI46325_C18_neg','QI42688_C18_neg','QI54634_C18_neg','QI45506_C18_neg','QI5007_C8_pos','QI14471_HILIC_pos',  'QI3365_C8_pos','QI5103_C8_pos','QI13024_C18_neg')]

mwas_new$Δ_intrx_lmm_fmla




```

# Comparisons of variants significant in overlapping categories
```{r}
dt <- data.table(
#signif_mvpa_proxy      = mwas[,          lmm_p < p_thresh_mwas],
#signif_SNP_assoc       = mwas[,     main_lmm_p < p_thresh_mwas],
##signif_mediator        = mwas[, med_lmm_acme_p < p_thresh_mwas],
#signif_GxM             = mwas[,    intrx_lmm_p < p_thresh_mwas],
#signif_nonlinear       = mwas[,       sq_lmm_p < p_thresh_mwas],
#
#signif_mediator_loose  = mwas[, med_lmm_acme_p < p_thresh_med_lmm  ]
#signif_SNP_assoc_loose = mwas[,     main_lmm_p < p_thresh_intrx_lmm],
#signif_GxM_loose       = mwas[,    intrx_lmm_p < p_thresh_intrx_lmm],
#signif_nonlinear_loose = mwas[,       sq_lmm_p < p_thresh_intrx_lmm]

#mediators_w_signif_SNP_assoc = mwas[med_lmm_acme_p < p_thresh_med_lmm,     main_lmm_p < p_thresh_intrx_lmm],
#mediators_w_signif_GxM       = mwas[med_lmm_acme_p < p_thresh_med_lmm,    intrx_lmm_p < p_thresh_intrx_lmm],
#mediators_w_signif_nonlinear = mwas[med_lmm_acme_p < p_thresh_med_lmm,       sq_lmm_p < p_thresh_intrx_lmm]

signif_mvpa_proxy_.01= mwas[,          lmm_p < 0.01],
signif_SNP_assoc_.01 = mwas[,     main_lmm_p < 0.01],
signif_GxM_.01       = mwas[,    intrx_lmm_p < 0.01],
signif_nonlinear_.01 = mwas[,       sq_lmm_p < 0.01],
signif_mediator_.01  = mwas[, med_lmm_acme_p < 0.01]
)
dt[is.na(dt)] <- F
dt[, names(.SD) := lapply(.SD,as.numeric), .SDcols=is.logical]

upset(dt)

sum(signif_mediator        & signif_GxM_loose, na.rm=T)
sum(signif_nonlinear_loose & signif_GxM_loose, na.rm=T)

which(signif_mediator        & signif_GxM_loose)
which(signif_nonlinear_loose & signif_GxM_loose)

sum(signif_mediator_loose, na.rm=T)


library(UpSetR)
movies <- read.csv( system.file("extdata", "movies.csv", package = "UpSetR"), header=T, sep=";" )
mutations <- read.csv( system.file("extdata", "mutations.csv", package = "UpSetR"), header=T, sep = ",")
upset(mutations)
movies

#cat(i<<-i+1,'/',.N,'\r') # Progress bar
```

