---
output: html_document
---
```{css, eval=T, echo=F}
pre code { overflow-x: scroll; white-space: pre; }
```
```{r setup, include=F}
# Uncomment to cheat and load pre-computed results to avoid waiting hours to regenerate an html render.
#knitr::opts_chunk$set(eval=F)
```
```{r cheat, eval=T, cache=T, include=F}
#load('medmod-250828.RData')
```

# Setup
```{r lib, eval=T, message=F}
#remotes::install_github("variani/lme4qtl")
library(data.table)
library(ggplot2)
library(ggrepel)
library(gtExtras)
library(lme4)
library(lme4qtl)
library(lmerTest)
library(lmtest)
library(mediation)
library(parallel)

options(mc.cores=4L)
options(datatable.na.strings=c('NA',''))

# Workaround so that lmerTest works on lme4qtl model objects
myLmer <- \(formula, data, relmat) {
  model <- relmatLmer(formula, data, relmat=relmat)
  lmerTest:::as_lmerModLT(model, as.function(model))
}
```
```{r data}
variants_of_interest <- fread(cmd='gcloud storage cat gs://fc-secure-4a392455-5587-4d6f-b8bd-01a1f834ae63/variants_of_interest.csv')
cpaids <- variants_of_interest$cpaid

data <- fread('analysis_df.csv')[
  ][, mesa_id    := as.factor(mesa_id)
  ][, exam       := as.factor(exam)
  # Variants are numeric (0/1/2), but we temporarily set to character so they don't get scale()'d.
  ][, names(.SD) := lapply(.SD, as.character), .SDcols=cpaids
  ][, names(.SD) := lapply(.SD, scale       ), .SDcols=is.numeric
  ][, names(.SD) := lapply(.SD, as.numeric  ), .SDcols=cpaids
]

data_exam1 <- data[exam==1]
data_exam2 <- data[exam==2]
data_exam3 <- data[exam==3]
data_exam4 <- data[exam==4]
data_exam5 <- data[exam==5]

load('mesa_km.RData') # Kinship (with ids remapped in script 01c)

numeric_colnms  <- names(data)[sapply(data,is.numeric)]
baseline_colnms <- paste0(numeric_colnms,'_0')
delta_colnms    <- paste0(numeric_colnms,'_Δ')
data[, (baseline_colnms) := .SD[exam==1],    .SDcols=numeric_colnms,     by='mesa_id'              ] |> invisible()
data[, (   delta_colnms) := data[,.SD,.SDcols=numeric_colnms] - data[,.SD,.SDcols=baseline_colnms] ] |> invisible()
#data[mesa_id %in% mesa_id[duplicated(mesa_id)], .SD, .SDcols=grepl('QI19074_HI|exam',names(data)), by='mesa_id'] # Sanity check

met_nms <- names(fread('metabolomics/QCd/merged_QCd.csv',nrows=0,drop=1))
covars <- c('site', 'sex', 'age', 'ses_score', 'drinks_per_week', 'smoking', 'ahei_score', paste0('gPC',1:5), paste0('gPC',1:5,'*mvpa_wins'), 'race')

calc_eff_n_metabolites <- \(met_nms) {
  met_nms <- met_nms[!is.na(met_nms)] |> unique()
  met_mtx <- data[ rowSums(is.na(data[,..met_nms]))==0, ..met_nms ] |> as.matrix(rownames=1) # Only samples having data for ALL metabolomics methods
  met_eigvals <- prcomp(met_mtx, scale=T, center=T)$sdev^2
  met_eff_n <- sum(met_eigvals)^2 / sum(met_eigvals^2)
}
```

# Define formulas
Main models of interest:

* __Metabolome-wide PA association analysis (MWAS)__: \
  `met ~ mvpa_wins + covars` \
* __HDL association analysis__: \
  `met ~ hdl_log + covars` \
* __Interaction Model__: \
  `hdl_log ~ met*SNP + sex*SNP + covars`
```{r fmlas}
runs <- CJ(met=met_nms, snp=cpaids, lm_or_lmm=c('LM','LMM'), exam=1,
  model=c('met~pa','met~hdl','met~snp',
          'hdl~met^2','hdl~met_RESET',
          'hdl~met*snp','hdl~Δmet*snp',

          'hdl~pa', 'hdl~pa_RESET', 'pa~snp',
          'hdl~pa*snp', 'hdl~pa*snp_no_sex*snp'))[

  # Exceptions
  ][!(lm_or_lmm=='LMM' & grepl('RESET',model))  # RESET models must be LMs
  ][!(lm_or_lmm=='LM'  & grepl(  'Δ'  ,model))  # Decomposed models must be LMMs
  ][!grepl('met',model), met := NA              # Rm per-met/per-snp rows for models w/o a met/snp term
  ][!grepl('snp',model), snp := NA
  ][lm_or_lmm=='LMM',    exam := NA
  ][!duplicated(paste(model,lm_or_lmm,exam,met,snp)) 


  # Basic associations
  ][model=='hdl~pa',  fmla := paste0( 'hdl_log', ' ~ ','mvpa_wins')
  ][model=='met~pa',  fmla := paste0(    met,    ' ~ ','mvpa_wins')
  ][model=='met~hdl', fmla := paste0(    met,    ' ~ ', 'hdl_log' )
  ][model=='met~snp', fmla := paste0(    met,    ' ~ ',    snp    )
  ][model=='pa~snp',  fmla := paste0('mvpa_wins',' ~ ',    snp    )

  # Nonlinearity
  ][model=='hdl~met^2',     fmla := paste0('hdl_log',' ~ ',    met,' + I(',met,'^2)')
  ][model=='hdl~met_RESET', fmla := paste0('hdl_log',' ~ ',    met                  )
  ][model=='hdl~pa_RESET',  fmla := paste0('hdl_log',' ~ ','mvpa_wins'              )

  # Interaction
  ][model=='hdl~pa*snp',            fmla := paste0('hdl_log',' ~ ','mvpa_wins','*',snp,' + ','sex','*',snp)
  ][model=='hdl~pa*snp_no_sex*snp', fmla := paste0('hdl_log',' ~ ','mvpa_wins','*',snp                    )
  ][model=='hdl~met*snp',           fmla := paste0('hdl_log',' ~ ',    met,    '*',snp,' + ','sex','*',snp)

  # hdl~met*snp model, except met term decomponsed into:        ↓baseline (exam1)   ↓delta (examX - exam1).
  ][model=='hdl~Δmet*snp', fmla := paste0('hdl_log',' ~ ',met,'_0','*',snp,' + ',met,'_Δ','*',snp,' + ','sex','*',snp)


  # Finalize
  ][, fmla := paste(fmla, '+', paste(covars,collapse='+'))           # Add covars common to all models
  ][model=='pa~snp', fmla := gsub('\\+gPC[0-9]\\*mvpa_wins','',fmla) #   (Except pa~snp shouldn't have pa itself in the covariates)
  ][lm_or_lmm=='LMM', fmla := paste(fmla, '+', '(1|mesa_id)')        # Add random intercept to all LMMs

  # Specify the term of interest whose stats will be extracted...
  # ...by specifying a pattern to grep (b/c interaction terms could be named either "met:snp" OR "snp:met").
  ][model %in% c('met~pa','hdl~pa'),                    term_pat := '^mvpa_wins$'
  ][model %in% c('met~hdl'),                            term_pat := '^hdl_log$'
  ][model %in% c('met~snp','pa~snp'),                   term_pat := paste0('^',snp,'$')
  ][model %in% c('hdl~met^2'),                          term_pat := paste0('I\\(',met,'\\^2\\)')
  ][model %in% c('hdl~met*snp'),                        term_pat := paste0(met,':',snp,'|',snp,':',met)
  ][model %in% c('hdl~Δmet*snp'),                       term_pat := paste0(met,'_Δ:',snp,'|',snp,':',met,'_Δ') # TODO extract baseline term too
  ][model %in% c('hdl~pa*snp','hdl~pa*snp_no_sex*snp'), term_pat := paste0('mvpa_wins:',snp,'|',snp,':mvpa_wins')
  ][is.na(term_pat), term_pat := 'NONE' # (grepping by NA would cause problems.)

] |> merge(by.x='met', by.y='unique_met_id', all=T, # Merge to metabolite metadata (M/Z, RT, common name, etc.)
           fread('metabolomics/met_info.csv')[unique_met_id %in% met_nms, .(unique_met_id,Method,MZ,RT,HMDB_ID,Metabolite)])
```

# Time to compute!
## Confirm HDL ~ PA assocation
Model: `log_hdl ~ mvpa_wins + covariates + (1|mesa_id)`
Only top 10 model terms are shown.
```{r confirm-hdl-pa, eval=T, echo=F}
runs[model=='hdl~pa' & lm_or_lmm=='LMM']$fmla |>
  myLmer(data, list(mesa_id=km)) |>
  summary() |> coef() |>
  as.data.table(keep.rownames='term') |>
  (\(dt) dt[order(`Pr(>|t|)`)])() |>
  head(10) |>
  gt() |>
  gt_highlight_rows(rows=term=='mvpa_wins')
```

## Confirm HDL ~ GxPA associations
Model: `log_hdl ~ snp*mvpa_wins + snp*sex + covars + (1|mesa_id)`
```{r confirm-hdl-gxpa, eval=T, echo=F}
runs[
  # Run models
  ][model=='hdl~pa*snp' & lm_or_lmm=='LM'
    , c('est','se',     't','p') := transpose(Map(fmla,term_pat, f=\(fmla,pat) {
        coefs <- summary(lm(fmla,data_exam1))$coefficients
        coefs[grepl(pat,rownames(coefs)),]
      }))
  ][model=='hdl~pa*snp' & lm_or_lmm=='LMM'
    , c('est','se','df','t','p') := transpose(Map(fmla,term_pat, f=\(fmla,pat) {
        coefs <- summary(myLmer(fmla,data,list(mesa_id=km)))$coefficients
        coefs[grepl(pat,rownames(coefs)),]
      }))

  # Display
  ][model=='hdl~pa*snp' & lm_or_lmm=='LMM'
  ][variants_of_interest, on=.(snp=cpaid)
  ][order(p)
  ][, .(snp,gene,rsid,analysis,annotation,GxPA_p=p)
] |> gt() |> gt_highlight_rows(rows=GxPA_p<0.05)
```

Moving foward with __only__ the two nomically-significant variants.
```{r subset-to-signif-hdl-gxpa}
signif_snps <- runs[model=='hdl~pa*snp' & lm_or_lmm=='LMM' & p < 0.05, snp]
runs <- runs[snp %in% signif_snps | is.na(snp)]
```

## Run all basic models
```{r run-all-basic-models, message=F}
i <- 0
runs <- runs[lm_or_lmm=='LM'
  , c('est','se',     't','p') := transpose(mcMap(fmla,term_pat, f=\(fmla,pat) {
      message('  ',(i<<-i+1)*getOption('mc.cores'),'/',.N,'\r',appendLF=F) # Progress bar
      s <- summary( lm(fmla, data_exam1) )                                 # Run model
      s$coefficients[grepl(pat,rownames(s$coefficients)),]                 # Extract results
}))]

i <- 0
runs <- runs[lm_or_lmm=='LMM'
  , c('est','se','df','t','p','singular_fit') := transpose(mcMap(fmla,term_pat, f=\(fmla,pat) {
      message('  ',(i<<-i+1)*getOption('mc.cores'),'/',.N,'\r',appendLF=F) # Progress bar
      s <- summary( myLmer(fmla, data, relmat=list(mesa_id=km)) )          # Run model
      c(s$coefficients[grepl(pat,rownames(s$coefficients)),],              # Extract results
        grepl('singular', s$optinfo$conv$lme4$messages)     )              # Extract singular fit warning msgs
}))]

i <- 0
runs <- runs[grepl('RESET',model)
  , p := transpose(mclapply(fmla, FUN=\(fmla) {
      message('  ',(i<<-i+1)*getOption('mc.cores'),'/',.N,'\r',appendLF=F) # Progress bar
      resettest(as.formula(fmla), data=data_exam1)$p.value                 # Run model & extract p
}))]
```

# Results summary
```{r get-signif-mets-and-thresholds, echo=F}
`p_thresh_mwas` <- 0.05/calc_eff_n_metabolites( unique(runs$met) )

`signif_mets_met~pa` <-
  runs[lm_or_lmm=='LMM', .(met = 
    .SD[model=='met~pa' & p < p_thresh_mwas, met])]

`p_thresh_after_met~pa` <- 0.05/calc_eff_n_metabolites(`signif_mets_met~pa`$met)

`signif_mets_met~pa->met~hdl` <-
  runs[lm_or_lmm=='LMM', .(met = intersect(`signif_mets_met~pa`$met,
    .SD[model=='met~hdl' & p < `p_thresh_after_met~pa`, met]))]

`p_thresh_after_met~pa->met~hdl` <- 0.05/calc_eff_n_metabolites(`signif_mets_met~pa->met~hdl`$met)

`signif_mets_met~pa->met~hdl->hdl~met*snp` <-
  runs[lm_or_lmm=='LMM', by=snp, .(met = intersect(`signif_mets_met~pa->met~hdl`$met,
    .SD[model=='hdl~met*snp' & p < `p_thresh_after_met~pa->met~hdl`, met]))]

`signif_mets_hdl~met*snp` <-
  runs[lm_or_lmm=='LMM', by=snp, .(met = 
    .SD[model=='hdl~met*snp' & p < p_thresh_mwas, met])]

#`p_thresh_after_met~pa->met~hdl->hdl~met*snp` <- 0.05/calc_eff_n_metabolites(`signif_mets_met~pa->met~hdl->hdl~met*snp`$met)
`p_thresh_after_met~pa->met~hdl->hdl~met*snp` <- `signif_mets_met~pa->met~hdl->hdl~met*snp`[, by=snp, .(p = 
  tryCatch(expr=0.05/calc_eff_n_metabolites(met), error=\(e)0.05))]
```

```{r, include=F}
#save.image('all_models_p_threshes-250814.RData')
```

```{r, include=F, eval=T}
# Because the variable names need backticks, hard to display in markdown. So defining new names here.
n_signif_met_pa <- length(`signif_mets_met~pa`$met)
n_signif_met_pa_met_hdl <- length(`signif_mets_met~pa->met~hdl`$met)
n_eff_mets_met_pa_met_hdl <- signif(0.05/`p_thresh_after_met~pa->met~hdl`,3)
```

Legend:

| Abbreviation | Result comes from which term in which model? | Description |
|:-------|:----------------------------------------|:-----------------|
| `hdl~met*snp`  | `hdl_log ~ `__`SNP*mvpa_wins`__` + SNP*sex + covars`                             | GxM interaction
| `hdl~Δmet*snp` | `hdl_log ~ `__`SNP*(met_exam1-met_examN)`__` + SNP*met_exam1 + SNP*sex + covars` | GxM interaction (decomposed)
| `met~snp`      | `met ~ `__`SNP`__` + covars`                                                     | Genetic main effect
| `hdl~met^2`    | `hdl_log ~ met + `__`met^2`__` + covars`                                         | Nonlinear effect
| `hdl~met_RESET`| `resettest(hdl_log ~ met_exam1 + covars)`                                        | RESET test p-value (exam1-only LM, not LMM)

## Genotype-independent models:
Total number of metabolites to start wtih: __`r length(met_nms)`__ (effective # of tests: `r signif(0.05/p_thresh_mwas,3)`).

| Model | Significant Metabolites |
|:------|:------------------------|
| `met ~ PA` MWAS                                                                 | `r n_signif_met_pa`         |
| `met ~ log(HDL)` (on `r n_signif_met_pa` significant metabolites from the MWAS) | `r n_signif_met_pa_met_hdl` |

## GxPA (analysis funnel)
Number of metabolites to start wtih: __`r n_signif_met_pa_met_hdl`__ (effective # of tests: `r n_eff_mets_met_pa_met_hdl`).

```{r, echo=F, eval=T}
runs[
  ][lm_or_lmm=='LMM'
  ][, .(Metabolite,met,snp,model,p=signif(p,3))
  ][`signif_mets_met~pa->met~hdl->hdl~met*snp`, on=.(met,snp)
  ][, dcast(.SD, value.var='p', Metabolite + met + snp ~ model)
  ][order(snp,`hdl~met*snp`)
  ][, `#` := 1:.N, by=snp
] |> gt() |> tab_header('Metabolite*SNP P-values', 'for metabolites significant from GxPA analysis funnel')

runs[
  ][lm_or_lmm=='LMM' | grepl('RESET',model)
  ][, .(Metabolite,met,snp,model,p=signif(p,3))
  ][is.na(snp) & met %in% `signif_mets_met~pa->met~hdl->hdl~met*snp`$met
    , dcast(.SD, value.var='p', Metabolite + met ~ model)
] |> gt() |> tab_header('SNP-independent P-values', 'for metabolites significant from GxPA analysis funnel')

```

## GxPA (MWIS)
Total number of metabolites to start wtih: __`r length(met_nms)`__ (effective # of tests: `r signif(0.05/p_thresh_mwas,3)`).

```{r, echo=F, eval=T}
runs[
  ][lm_or_lmm=='LMM'
  ][, .(Metabolite,met,snp,model,p=signif(p,3))
  ][`signif_mets_hdl~met*snp`, on=.(met,snp)
  ][, dcast(.SD, value.var='p', Metabolite + met + snp ~ model)
  ][order(snp,`hdl~met*snp`)
  ][, `#` := 1:.N, by=snp
] |> gt() |> tab_header('Metabolite*SNP P-values', 'for metabolites significant from GxPA MWIS')

runs[
  ][lm_or_lmm=='LMM' | grepl('RESET',model)
  ][, .(Metabolite,met,snp,model,p=signif(p,3))
  ][is.na(snp) & met %in% `signif_mets_hdl~met*snp`$met
    , dcast(.SD, value.var='p', Metabolite + met ~ model)
] |> gt() |> tab_header('SNP-independent P-values', 'for metabolites significant from GxPA MWIS')
```

## Write
```{r eval=F}
fwrite(runs,'runs.csv')
```

## Mediated moderation
```{r}
i <- 0
tmp <- funion(`signif_mets_met~pa->met~hdl->hdl~met*snp`, `signif_mets_hdl~met*snp`)[
  # Calculate the minimum number of sims needed to pass the p-value threshold, plus a little 20% extra.
  ][, by=c(snp2='snp')
    , n_sims := `p_thresh_after_met~pa->met~hdl->hdl~met*snp`[snp==snp2, ceiling(1.2/p)]

  # mediate() errored when I used a*b interaction terms in these models.
  # As a workaround I manually construct 'axb' product terms and use those instead.
  # See LabArchives days 250827-28 for my troubleshooting, but I never got to the bottom of it.
  ][, c('treat','mediator') := transpose(Map(snp,met, f=\(snp,met) {
        treat    <- paste0(snp,'x','mvpa_wins')
        mediator <- paste0(snp,'x',    met    )
        data[, (treat)    := .SD[[1]]*.SD[[2]], .SDcols=c(snp,'mvpa_wins')]
        data[, (mediator) := .SD[[1]]*.SD[[2]], .SDcols=c(snp,    met    )]
        list(treat,mediator)
      }))

  # Since we're not using 'a*b' formula syntax, have to manually add + a and + b terms.
  ][, fmla_mediator := paste0(snp,'x',met,' ~ ',snp,'x','mvpa_wins',' + ',met,' + ',snp,' + ','mvpa_wins'                  )
  ][, fmla_outcome  := paste0( 'hdl_log' ,' ~ ',snp,'x','mvpa_wins',' + ',met,' + ',snp,' + ','mvpa_wins',' + ',snp,'x',met)

  ## Finalize
  ][, fmla_mediator := paste(fmla_mediator, '+', paste(covars,collapse='+')) # Add covars
  ][, fmla_outcome  := paste(fmla_outcome,  '+', paste(covars,collapse='+'))
  ][, fmla_mediator := paste(fmla_mediator, '+', '(1|mesa_id)')              # Add random intercept
  ][, fmla_outcome  := paste(fmla_outcome,  '+', '(1|mesa_id)')

  ][, c('ACME_p','mediation_prop','N') := transpose(mcMap(fmla_mediator,fmla_outcome,treat,mediator,n_sims, f=\(fmla_m,fmla_y,treat,mediator,n_sims) {
        message('  ',(i<<-i+1)*getOption('mc.cores'),'/',.N,'\r',appendLF=F) # Progress bar
        mediate(
          myLmer(fmla_m, data[!is.na(hdl_log)], relmat=list(mesa_id=km)) |> `class<-`('lmerMod'),
          myLmer(fmla_y, data                 , relmat=list(mesa_id=km)) |> `class<-`('lmerMod'),
          treat=treat, mediator=mediator, robustSE=T,
          sims=n_sims
        )[c('d0.p','n0','nobs')] # Extract only the ACME p-val, mediation proportion, and N.
      }))
]

save.image('medmod-250828.RData')
```

```{r, echo=F, eval=T}
tmp[
  ][, .(met,snp,ACME_p,mediation_prop,N)
  ][order(snp,as.numeric(ACME_p))
  ][, `#` := 1:.N, by=snp
] |> gt() |> tab_header('Mediated Moderation Results', 'for metabolites significant from either GxPA MWIS or analysis funnel (the funnel ones are highlighted)') |>
  gt_highlight_rows(rows = met %in% `signif_mets_met~pa->met~hdl->hdl~met*snp`$met)
```




<!---

## Correlations
For metabolite identity exploration.
```{r eval=F}
signif_or_known_met_nms <- mwas[intrx_lmm_p_chr14_88305056_G_A  < p_thresh_mwas | !is.na(Metabolite), unique_met_id]
signif_met_nms <- mwas[intrx_lmm_p_chr14_88305056_G_A  < p_thresh_mwas, unique_met_id]
tmp <- data[, cor(.SD, method='pearson', use='complete.obs'), .SDcols=signif_or_known_met_nms]
tmp2 <- as.data.table(tmp, keep.rownames='unique_met_id')
tmp2 <- tmp2[, .SD, .SDcols=c('unique_met_id',signif_met_nms)]
colnames(tmp2) <- paste0('cor_',colnames(tmp2))
mwas_w_cors <- merge(mwas,tmp2,by.x='unique_met_id',by.y='cor_unique_met_id',all.x=T)
fwrite(mwas_w_cors,'mwas_w_cors.csv')

cats <- fread('metabolites_subclasses.csv')

library(ComplexHeatmap)
library(circlize)
Heatmap(
  cor_mtx <- tmp2[,-1],
  col=colorRamp2(c(-1,0, 1), c('blue','white','red')),
  #cell_fun = \(j,i,x,y,w,h,col) grid.text(cor_mtx[i,j]|>round(digits=2),x,y,gp=gpar(fontsize=10)),
  #show_heatmap_legend=F,
  column_title='Pearson correlations'
)
```

## GxE models:
```{r tbls2, eval=F, echo=F, results='asis'}
for(cpaid in cpaids) {
  cat('### ',cpaid,'\n')
  v <- list(intrx_lm_p=NA, intrx_lmm_p=NA, modmed_lm_acme_p=NA, p_thresh_modmed_lm=NA, p_thresh_modmed_lmm=NA)
  lapply(names(v), \(nm) mwas[, v[nm] <<- .SD, .SDcols=paste0(nm,'_',cpaid)])

  cat(paste('\n\n','Significant metabolites, GxE   MWAS  ',cpaid,'(LM) :', sum(na.rm=T, v$intrx_lm_p  < mwas$p_thresh_intrx_lm                                             ) ))
  cat(paste('\n\n','Significant metabolites, GxE "funnel"',cpaid,'(LM) :', sum(na.rm=T, v$intrx_lm_p  < mwas$p_thresh_intrx_lm & mwas$med_lm_acme_p  < mwas$p_thresh_med_lm) ))
  print(knitr::kable(mwas[, .SDcols = c('unique_met_id', paste0(c('intrx_lm_est_', 'intrx_lm_se_', 'intrx_lm_p_'),cpaid), 'p_thresh_intrx_lm'), .SD][ order(mwas[, .SD[[1]], .SDcols=paste0('intrx_lm_p_', cpaid)]) ][1:3] ))

  cat(paste('\n\n','Significant metabolites,   GxE    MWAS  ',cpaid,'(LMM):', sum(na.rm=T, v$intrx_lmm_p < mwas$p_thresh_intrx_lmm                                              ) ))
  cat(paste('\n\n','Significant metabolites,   GxE  "funnel"',cpaid,'(LMM):', sum(na.rm=T, v$intrx_lmm_p < mwas$p_thresh_intrx_lmm & mwas$med_lmm_acme_p < mwas$p_thresh_med_lmm) ))
  print(knitr::kable(mwas[, .SDcols = c('unique_met_id', paste0(c('intrx_lmm_est_', 'intrx_lmm_se_', 'intrx_lmm_p_'),cpaid), 'p_thresh_intrx_lmm'), .SD][ order(mwas[, .SD[[1]], .SDcols=paste0('intrx_lmm_p_',cpaid)]) ][1:3] ))

  cat(paste('\n\n','Significant metabolites, modmed "funnel"',cpaid,'(LM) :', sum(na.rm=T, v$modmed_lm_acme_p < v$p_thresh_modmed_lm  & mwas$med_lm_acme_p  < mwas$p_thresh_med_lm ) ))
  cat(paste('\n\n','Significant metabolites, modmed "funnel"',cpaid,'(LMM):', sum(na.rm=T, v$modmed_lm_acme_p < v$p_thresh_modmed_lmm & mwas$med_lmm_acme_p < mwas$p_thresh_med_lmm) ))
  print(knitr::kable(mwas[, .SDcols = c('unique_met_id', paste0(c('modmed_lm_acme_p_', 'p_thresh_modmed_lm_'),cpaid)), .SD][ order(mwas[, .SD[[1]], .SDcols=paste0('modmed_lm_acme_p_',cpaid)]) ][1:3] ))
}
```
--->





<!-- # Junk
```{r, include=F}
#load('250112_modmed_lmm_done.RData')
#load('250224-modemed_lmm_done-delta_cols.RData')
#load('250227-mwas_mainG-mwas_sq.RData')
load('250320.RData')

{ mwas[      lm_p < .05, plot(      lm_est/      lm_se,       lmm_est/      lmm_se, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM MWAS Z-scores\n       (metabolites with MWAS LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }
{ mwas[intrx_lm_p < .05, plot(intrx_lm_est/intrx_lm_se, intrx_lmm_est/intrx_lmm_se, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM Interaction Z-scores\n(metabolites with GxM  LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }
#mwas[lm_p < .1 & lm_est>0, boxplot(lm_est/lm_se, lmm_est/lmm_se, names=c('LM','LMM'), ylab='Z-score', main='Metabolites whose LM p-value was < 0.1\nPOSITIVE Z-scores')]
#mwas[lm_p < .1 & lm_est<0, boxplot(lm_est/lm_se, lmm_est/lmm_se, names=c('LM','LMM'), ylab='Z-score', main='Metabolites whose LM p-value was < 0.1\nNEGATIVE Z-scores')]
{ mwas[lm_p < .05, boxplot(abs(      lm_est/      lm_se), abs(      lmm_est/      lmm_se), names=c('LM','LMM'), ylab='|Z-score|', main='Metabolites whose LM p-value was < 0.1')] }
{ mwas[lm_p < .05, boxplot(abs(intrx_lm_est/intrx_lm_se), abs(intrx_lmm_est/intrx_lmm_se), names=c('LM','LMM'), ylab='|Z-score|', main='Metabolites whose **Interaction** LM p-value was < 0.1')] }
#mwas[                                  , plot(-log10(med_lm_acme_p+1e-24), -log10(med_lmm_acme_p+1e-24), col=rgb(0,0,0,0.2),pch=16)] + abline(0,1,col='red')
#mwas[                                  , plot(-log10(fifelse(is.na(med_lm_acme_p), 1,med_lm_acme_p +1e-24)),
#                                              -log10(fifelse(is.na(med_lmm_acme_p),1,med_lmm_acme_p+1e-24)),
#                                              col=rgb(0,0,0,0.2),pch=16)] + abline(0,1,col='red')

#cat(mwas[order(      lm_p)[1:30],       lm_p/      lmm_p],sep='\n')
#cat(mwas[order(intrx_lm_p)[1:30], intrx_lm_p/intrx_lmm_p],sep='\n')

{ mwas[      lm_p<.05, hist(log(abs(      lmm_est/      lmm_se)/abs(      lm_est/      lm_se)), xlab='log(LMM_Z / LM_Z)', breaks=40L, main='Metabolites whose LM p-value was < 0.1\nLog LMM/LM Z-score ratios') ]; abline(0,1e10,col='red') }
{ mwas[intrx_lm_p<.05, hist(log(abs(intrx_lmm_est/intrx_lmm_se)/abs(intrx_lm_est/intrx_lm_se)), xlab='log(LMM_Z / LM_Z)', breaks=40L, main='Metabolites whose **Interaction** LM p-value was < 0.1\nLog Interaction LMM/LM Z-score ratios') ]; abline(0,1e24,col='red')}

# Note that in later steps, the increased number of significant mets may just be because more were tested because they passed the threshold of the previous step.
# Although the p thresholds do control for the # of effective metabolites.
mwas[                                                                                            , sum(       lm_p       < p_thresh_mwas     )]
mwas[lm_p  < p_thresh_mwas                                                                       , sum(   med_lm_acme_p  < p_thresh_med_lm   )]
mwas[lm_p  < p_thresh_mwas & med_lm_acme_p  < p_thresh_med_lm                                    , sum( intrx_lm_p       < p_thresh_intrx_lm )]
mwas[lm_p  < p_thresh_mwas & med_lm_acme_p  < p_thresh_med_lm  & intrx_lm_p  < p_thresh_intrx_lm , sum(modmed_lm_acme_p  < p_thresh_modmed_lm)]
mwas[                                                                                            , sum(       lmm_p      < p_thresh_mwas     )]
mwas[lmm_p < p_thresh_mwas                                                                       , sum(   med_lmm_acme_p < p_thresh_med_lmm  )]
mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm                                   , sum( intrx_lmm_p      < p_thresh_intrx_lmm)]
mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm & intrx_lmm_p < p_thresh_intrx_lmm, sum(modmed_lm_acme_p  < p_thresh_modmed_lm)]
mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm & intrx_lmm_p < p_thresh_intrx_lmm, sum(modmed_lm_acme_p  < .1                )]

mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm & intrx_lmm_p < p_thresh_intrx_lmm, .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')]
mwas[lmm_p < p_thresh_mwas & med_lmm_acme_p < p_thresh_med_lmm                                   , .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')][order()]
mwas[lmm_p < p_thresh_mwas                                                                       , .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')]

# If we pretend we meant to do an interaction mwas
mwas[                                                                                            , sum( intrx_lm_p       < p_thresh_mwas     )]
mwas[                                                                                            , sum( intrx_lmm_p      < p_thresh_mwas     )]
mwas[intrx_lm_p  < p_thresh_mwas                                                                 , .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')]
mwas[intrx_lmm_p < p_thresh_mwas                                                                 , .SD, .SDcols=patterns('Met|met|MZ|RT|_est|_se|_p')][order(intrx_lm_p)]

#tmp <- mwas[intrx_lmm_p < p_thresh_mwas                                                                 , .SD, .SDcols=patterns('met|_est|_p')][order(intrx_lm_p)]
"%ni%" <- Negate("%in%")
tmp <- mwas[order(unique_met_id %ni% c('QI48826_C18_neg','QI7292_C8_pos','QI46325_C18_neg','QI42688_C18_neg','QI54634_C18_neg','QI45506_C18_neg','QI5007_C8_pos','QI14471_HILIC_pos',  'QI3365_C8_pos','QI5103_C8_pos','QI13024_C18_neg'))]
```
```{r}
# TODO: knitr tables like this: knitr::kable(mwas[lm_p   <p_thresh_mwas, .(MZ,RT,HMDB_ID,Metabolite,unique_met_id,lm_p   )][order(lm_p   )] )
```

## QQ
```{r qq}
mwas[
  ][ order(lmm_p)
  ][, nlp := -log10(lmm_p)
  ][, nlp_expect := -log10(ppoints(.N))
] |>
  ggplot(aes(x=nlp_expect, y=nlp)) +
    geom_point() +
    geom_abline(slope=1, intercept=0) +
    theme_bw()

# Genomic inflation λ
calc_λ  <- function(x, p=0.5){
  x <- x[!is.na(x)]
  x.quantile <- quantile(x, p)
  round(qchisq(1 - x.quantile, 1) / qchisq(p, 1), 2)
}
mwas[, calc_λ(lmm_p)]
```

## Volcano
```{r volc, warning=F, fig.width=12, fig.height=8}
ggplot(mwas, aes(x=lmm_est, y=-log10(lmm_p), color=lmm_p<p_thresh_mwas)) +
  geom_hline(yintercept=-log10(p_thresh_mwas), linetype="dashed") +
  geom_point(alpha=0.4, shape=16) +
  geom_text_repel(data=mwas[lmm_p==p_thresh_mwas], aes(label=Metabolite), size=3, show.legend=F) +
  scale_color_manual(name="Significant?\n(p<0.05/n_eff_metabolites)",values=c("FALSE"="black","TRUE"="red")) +
  guides(color = guide_legend(override.aes = list(size=5)))
```


```{r}
mwas[unique_met_id%in%c('QI3365_C8_pos', 'QI5103_C8_pos', 'QI13024_C18_neg', 'QI48826_C18_neg'),
     .(unique_met_id,MZ,RT,Metabolite,lm_p,lmm_p,intrx_lmm_p,intrx_lmm_p,med_lmm_prop)
]


cor(mwas$intrx_lmm_est,mwas$intrx_lmm_est_0,use='complete.obs')
cor(mwas$intrx_lmm_est,mwas$intrx_lmm_est_Δ,use='complete.obs')
cor(abs(mwas_new$intrx_lmm_est),abs(mwas_new$intrx_lmm_est_0),use='complete.obs')
cor(abs(mwas_new$intrx_lmm_est),abs(mwas_new$intrx_lmm_est_Δ),use='complete.obs')
ggplot(mwas_new, aes(x=intrx_lmm_est/intrx_lmm_se,y=intrx_lmm_est_0/intrx_lmm_se_0)) + geom_point() + geom_text_repel(data=mwas_new, aes(label=unique_met_id), size=3, show.legend=F) + geom_abline(slope=1)
ggplot(mwas_new, aes(x=intrx_lmm_est/intrx_lmm_se,y=intrx_lmm_est_Δ/intrx_lmm_se_Δ)) + geom_point() + geom_text_repel(data=mwas_new, aes(label=unique_met_id), size=3, show.legend=F) + geom_abline(slope=1)

ggplot(mwas_new, aes(x=intrx_lmm_est_0/intrx_lmm_se_0,y=intrx_lmm_est_Δ/intrx_lmm_se_Δ)) + geom_point() + geom_text_repel(data=mwas, aes(label=unique_met_id), size=3, show.legend=F) + geom_abline(slope=1)

plot(abs(mwas$intrx_lmm_est),abs(mwas$intrx_lmm_est_0), main='baseline')
plot(abs(mwas$intrx_lmm_est),abs(mwas$intrx_lmm_est_Δ), main='delta')
#unique_met_id %in% c('QI3365_C8_pos', 'QI5103_C8_pos', 'QI13024_C18_neg', 'QI48826_C18_neg', 'QI7292_C8_pos', 'QI46325_C18_neg', 'QI42688_C18_neg', 'QI54634_C18_neg', 'QI45506_C18_neg', 'QI5007_C8_pos', 'QI14471_HILIC_pos')



{ mwas[unique_met_id %in% mets, plot(      lm_est/      lm_se,       lmm_est/      lmm_se, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM MWAS Z-scores\n       (metabolites with MWAS LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }
{ mwas[unique_met_id %in% mets, plot(intrx_lm_est/intrx_lm_se, intrx_lmm_est/intrx_lmm_se, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM Interaction Z-scores\n(metabolites with GxM  LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }

{ mwas[unique_met_id %in% mets, plot(      lm_est,       lmm_est, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM MWAS Z-scores\n       (metabolites with MWAS LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }
{ mwas[unique_met_id %in% mets, plot(intrx_lm_est, intrx_lmm_est, col=rgb(0,0,0,0.2),pch=16,main='LM vs. LMM Interaction Z-scores\n(metabolites with GxM  LM p<0.1)',xlab='LM Z',ylab='LMM Z')] + abline(0,1,col='red') + abline(0,999) + abline(0,0) }


# df w/ ests and z for LM & LMM, decomposed lmm baseline and delta... 
# Plot w/ the 11 mets as the x axis, and above each is a colored dot for each of the above stats.
# will prolly ned long format
mets <- c('QI48826_C18_neg','QI7292_C8_pos','QI46325_C18_neg','QI42688_C18_neg','QI54634_C18_neg','QI45506_C18_neg','QI5007_C8_pos','QI14471_HILIC_pos',  'QI3365_C8_pos','QI5103_C8_pos','QI13024_C18_neg')
df <- mwas[
  ][ unique_met_id %in% mets
    , .(unique_met_id,
        intrx_lm_est,
        intrx_lmm_est_0,
        intrx_lmm_est,
        intrx_lmm_est_Δ,
        intrx_lm_z    = intrx_lm_est   /intrx_lm_se   ,
        intrx_lmm_z_0 = intrx_lmm_est_0/intrx_lmm_se_0,
        intrx_lmm_z   = intrx_lmm_est  /intrx_lmm_se  ,
        intrx_lmm_z_Δ = intrx_lmm_est_Δ/intrx_lmm_se_Δ
      )
] #|>
  melt(id.vars='unique_met_id')

ggplot(df[grepl('z',variable)], aes(x=unique_met_id,y=value,fill=variable)) + geom_bar(stat='identity',position='dodge') + theme(axis.text.x=element_text(angle=45, hjust = 1))
ggplot(df[grepl('est',variable)], aes(x=unique_met_id,y=value,fill=variable)) + geom_bar(stat='identity',position='dodge') + theme(axis.text.x=element_text(angle=45, hjust = 1))


ggplot(mwas_new, aes(x=unique_met_id,y=intrx_lm_est)) + geom_bar(stat='identity',position='dodge')

df[,unique_met_id:=NULL]
df[, names(.SD) := NULL, .SDcols=patterns('z')]
setnames(df,c('LM estimates','LMM estimates (baseline)','LMM estimates','LMM estimates (delta)'))#, 'LM z-scores','LMM z-scores (baseline)','LMM z-scores','LMM z-scores (delta)'))
Heatmap(
  cor_mtx <- cor(df,method='pearson'),
  col=colorRamp2(c(0.5, 1), c("white", "red")),
  cell_fun = \(j,i,x,y,w,h,col) grid.text(cor_mtx[i,j]|>round(digits=2),x,y,gp=gpar(fontsize=10)),
  show_heatmap_legend=F,
  column_title='Pearson correlations'
)




mwas_old <- copy(mwas)

mwas_new <- mwas[unique_met_id %in%  c('QI48826_C18_neg','QI7292_C8_pos','QI46325_C18_neg','QI42688_C18_neg','QI54634_C18_neg','QI45506_C18_neg','QI5007_C8_pos','QI14471_HILIC_pos',  'QI3365_C8_pos','QI5103_C8_pos','QI13024_C18_neg')]

mwas_new$Δ_intrx_lmm_fmla




```

# Comparisons of variants significant in overlapping categories
```{r}
dt <- data.table(
#signif_mvpa_proxy      = mwas[,          lmm_p < p_thresh_mwas],
#signif_SNP_assoc       = mwas[,     main_lmm_p < p_thresh_mwas],
##signif_mediator        = mwas[, med_lmm_acme_p < p_thresh_mwas],
#signif_GxM             = mwas[,    intrx_lmm_p < p_thresh_mwas],
#signif_nonlinear       = mwas[,       sq_lmm_p < p_thresh_mwas],
#
#signif_mediator_loose  = mwas[, med_lmm_acme_p < p_thresh_med_lmm  ]
#signif_SNP_assoc_loose = mwas[,     main_lmm_p < p_thresh_intrx_lmm],
#signif_GxM_loose       = mwas[,    intrx_lmm_p < p_thresh_intrx_lmm],
#signif_nonlinear_loose = mwas[,       sq_lmm_p < p_thresh_intrx_lmm]

#mediators_w_signif_SNP_assoc = mwas[med_lmm_acme_p < p_thresh_med_lmm,     main_lmm_p < p_thresh_intrx_lmm],
#mediators_w_signif_GxM       = mwas[med_lmm_acme_p < p_thresh_med_lmm,    intrx_lmm_p < p_thresh_intrx_lmm],
#mediators_w_signif_nonlinear = mwas[med_lmm_acme_p < p_thresh_med_lmm,       sq_lmm_p < p_thresh_intrx_lmm]

signif_mvpa_proxy_.01= mwas[,          lmm_p < 0.01],
signif_SNP_assoc_.01 = mwas[,     main_lmm_p < 0.01],
signif_GxM_.01       = mwas[,    intrx_lmm_p < 0.01],
signif_nonlinear_.01 = mwas[,       sq_lmm_p < 0.01],
signif_mediator_.01  = mwas[, med_lmm_acme_p < 0.01]
)
dt[is.na(dt)] <- F
dt[, names(.SD) := lapply(.SD,as.numeric), .SDcols=is.logical]

upset(dt)

sum(signif_mediator        & signif_GxM_loose, na.rm=T)
sum(signif_nonlinear_loose & signif_GxM_loose, na.rm=T)

which(signif_mediator        & signif_GxM_loose)
which(signif_nonlinear_loose & signif_GxM_loose)

sum(signif_mediator_loose, na.rm=T)


library(UpSetR)
movies <- read.csv( system.file("extdata", "movies.csv", package = "UpSetR"), header=T, sep=";" )
mutations <- read.csv( system.file("extdata", "mutations.csv", package = "UpSetR"), header=T, sep = ",")
upset(mutations)
movies

#cat(i<<-i+1,'/',.N,'\r') # Progress bar
```


-->
