```{css echo=F}
pre {
  overflow-x: scroll;
}
pre code {
  white-space: pre;
}
```
# Setup
```{r setup, message=F}
library(data.table)
library(ggplot2)
library(ggrepel)
library(mediation)
library(parallel)

options(datatable.na.strings=c('NA',''))
data <- fread('analysis_df.csv')[metabolomics_exam==1]
met_ids <- names(fread('metabolomics/QCd/merged_QCd.csv',nrows=0,drop=1))
covars <- c('site', 'sex', 'age', 'ses_score', 'income', 'drinks_per_week', 'smoking', 'ahei_score', 'dash_score', paste0('gPC', 1:5), 'race')

# Data table that will will gradually fill with the results of each analysis
mwas <- fread('metabolomics/met_info.csv')[unique_met_id %in% met_ids, .(HMDB_ID,Metabolite,unique_met_id)]

calc_eff_n_metabolites <- \(met_ids) {
  met_mtx <- data[ rowSums(is.na(data[,..met_ids]))==0, ..met_ids ] |> as.matrix(rownames=1) # Only samples with data for ALL metabolomics methods
  met_eigvals <- prcomp(met_mtx, scale=T, center=T)$sdev^2
  met_eff_n <- sum(met_eigvals)^2 / sum(met_eigvals^2)
}
```

# MWAS on genotypes and exposures
Model: `lm(met ~ [snp or exposure] + covars)`
```{r mwas, cache=T, cache.depend=digest::digest(mwas), results='hide'}
p_thresh_mwas <- 0.05/calc_eff_n_metabolites(mwas[,unique_met_id])

mwas <- mwas[
  # Run linear models
  ][, results := mapply(unique_met_id,.I, FUN = \(met,i) {
        cat(i,'/',.N,'\r') # Progress bar
        fmla <- paste(met,'~ mvpa_wins +',paste(covars,collapse='+'))
        model <- lm(fmla,data)
        coefs <- summary(model)$coefficients
        coefs[rownames(coefs)=='mvpa_wins',]
      }, SIMPLIFY=F)

  # Extract results we want
  ][, mwas_estimate := sapply(results, '[', 'Estimate'  )
  ][, mwas_se       := sapply(results, '[', 'Std. Error')
  ][, mwas_p        := sapply(results, '[', 'Pr(>|t|)'  )
  ][, mwas_q        := p.adjust(mwas_p, method='BH')
  ][, mwas_signif   := mwas_p < p_thresh_mwas

  # Cleanup
  ][, results := NULL
]
```
```{r}
knitr::kable(mwas[mwas_signif==T, .(HMDB_ID,Metabolite,unique_met_id,mwas_p)][order(mwas_p)] )
```

## QQ
```{r qq}
mwas[
  ][ order(mwas_p)
  ][, nlp := -log10(mwas_p)
  ][, nlp_expect := -log10(ppoints(.N))
] |>
  ggplot(aes(x=nlp_expect, y=nlp)) +
    geom_point() +
    geom_abline(slope=1, intercept=0) +
    theme_bw()

# Genomic inflation λ
calc_λ  <- function(x, p=0.5){
  x <- x[!is.na(x)]
  x.quantile <- quantile(x, p)
  round(qchisq(1 - x.quantile, 1) / qchisq(p, 1), 2)
}
mwas[, calc_λ(mwas_p)]
```

## Volcano
```{r volc, warning=F, fig.width=16, fig.height=12}
ggplot(mwas, aes(x=mwas_estimate, y=-log10(mwas_p), color=mwas_p<p_thresh_mwas)) +
  geom_hline(yintercept=-log10(p_thresh_mwas), linetype="dashed") +
  geom_point(alpha=0.4, shape=16) +
  geom_text_repel(data=mwas[mwas_signif==T], aes(label=Metabolite), size=3, show.legend=F) +
  scale_color_manual(name="Significant?\n(p<0.05/n_eff_metabolites)",values=c("FALSE"="black","TRUE"="red")) +
  guides(color = guide_legend(override.aes = list(size=5)))
```

# Mediation Analysis
Comparing models\
`lm(  met   ~       mvpa_wins + covars)` and\
`lm(hdl_log ~ met + mvpa_wins + covars)` to find metabolites which mediate the hdl ~ pa relationship.
```{r mediation, cache=T, cache.depends=digest::digest(mwas), results='hide'}
p_thresh_med <- 0.05/calc_eff_n_metabolites(mwas[mwas_signif==T,unique_met_id])

mwas <- mwas[
  ][ mwas_signif==T
    , results := mapply(unique_met_id,.I, FUN = \(met,i) {
        cat(i,'/',.N,'\r') # Progress bar
        data <- data[!is.na(hdl_log)]
        fmla_m <- paste(   met,' ~ ',    '+ mvpa_wins +',paste(covars,collapse='+'))
        fmla_y <- paste('hdl_log ~ ',met,'+ mvpa_wins +',paste(covars,collapse='+'))
        mediate(
          lm(fmla_m,data),
          lm(fmla_y,data),
          treat='mvpa_wins', mediator=met,
          robustSE=T, sims=500 # Enough sims to reach the p threshold
        )[c('d0.p', 'n0', 'nobs')] # Select only the results we need. Otherwise eats up a lot of memory.
      }, SIMPLIFY=F)

  # Extract results we want
  ][lengths(results)>0, med_acme_p    := sapply(results,'[[','d0.p')
  ][lengths(results)>0, med_prop      := sapply(results,'[[','n0'  )
  ][lengths(results)>0, med_n_samples := sapply(results,'[[','nobs')
  ][lengths(results)>0, med_signif    := med_acme_p < p_thresh_med

  ## Cleanup
  ][, results := NULL
]
```
```{r mediation_tbl}
knitr::kable(mwas[med_signif==T,.(HMDB_ID,Metabolite,unique_met_id,med_acme_p,med_prop,med_n_samples)][order(med_acme_p)] )
```

# Interaction analysis
Model: `lm(hdl ~ met*rs295849 + covars)`.
```{r intrx, cache=T, cache.depends=digest::digest(mwas), results='hide'}
p_thresh_intrx <- 0.05/calc_eff_n_metabolites(mwas[med_signif==T,unique_met_id])

mwas <- mwas[
  ][ med_signif==T
    , results := mapply(unique_met_id,.I, FUN = \(met,i) {
        cat(i,'/',.N,'\r') # Progress bar
        #fmla <- paste0('hdl_log ~ ',met,'*rs295849 + ',paste0(covars,'*rs295849',collapse=' + '))
        fmla <- paste0('hdl_log ~ ',met,'*rs295849 + ',paste0(covars,            collapse=' + '))
        model <- lm(fmla,data)
        coefs <- summary(model)$coefficients
        coefs[grepl(met,rownames(coefs)) &
              grepl(':',rownames(coefs))  ,]
      }, SIMPLIFY=F)

  ][lengths(results)>0, intrx_estimate := sapply(results, '[', 'Estimate'  )
  ][lengths(results)>0, intrx_se       := sapply(results, '[', 'Std. Error')
  ][lengths(results)>0, intrx_p        := sapply(results, '[', 'Pr(>|t|)'  )
  ][lengths(results)>0, intrx_q        := p.adjust(intrx_p, method='BH')
  ][lengths(results)>0, intrx_signif   := intrx_p < p_thresh_intrx

  # Cleanup
  ][, results := NULL
]
```
```{r intrx_tbl}
#knitr::kable(mwas[intrx_signif==T][order(intrx_p)] ) # None signif :(
knitr::kable(mwas[intrx_p<0.1, .(HMDB_ID,Metabolite,unique_met_id,intrx_p)][order(intrx_p)] )
```

# Moderated Mediation
Comparing models\
`lm(  met   ~                mvpa_wins*rs295849 + covars)` and\
`lm(hdl_log ~ met*rs295849 + mvpa_wins*rs295849 + covars)`.
```{r moderated, cache=T, cache.depends=digest::digest(mwas), results='hide'}
#p_thresh_modmed <- 0.05/calc_eff_n_metabolites(mwas[intrx_signif==T,unique_met_id]) # None

mwas <- mwas[
  ][ intrx_p < 0.1
  #][ intrx_signif==T
    , results := mapply(unique_met_id,.I, FUN = \(met,i) {
        cat(i,'/',.N,'\r') # Progress bar
        data <- data[!is.na(hdl_log)]
        fmla_m <- paste(   met,   '~',                 'mvpa_wins * rs295849 +',paste(covars,collapse='+'))
        fmla_y <- paste('hdl_log','~',met,'* rs295849 + mvpa_wins * rs295849 +',paste(covars,collapse='+'))
        #fmla_m <- paste(   met,   '~',                 'mvpa_wins * rs295849 + age + sex')
        #fmla_y <- paste('hdl_log','~',met,'* rs295849 + mvpa_wins * rs295849 + age + sex')
        mediate(
          lm(fmla_m,data),
          lm(fmla_y,data),
          treat='mvpa_wins', mediator=met,
          robustSE=TRUE, sims=10 # N_sims doesn't need to be high for this step
        ) |>
        test.modmed(list(rs295849=0), list(rs295849=1), sims=100)
      }, SIMPLIFY=F)

  # Extract results we want
  ][lengths(results)>0, modmed_acme_p  := sapply(results, \(res) res[[1]]$p.value                                            )
  #][lengths(results)>0, modmed_acme_ci := sapply(results, \(res) res[[1]]$conf.int |> (\(x) paste0('(',x[1],',',x[2],')'))() )
  ][lengths(results)>0, modmed_ade_p   := sapply(results, \(res) res[[2]]$p.value                                            )
  #][lengths(results)>0, modmed_ade_ci  := sapply(results, \(res) res[[2]]$conf.int |> (\(x) paste0('(',x[1],',',x[2],')'))() )

  # Cleanup
  ][, results := NULL
]
```
```{r modmed-tbl}
#knitr::kable(mwas[modmed_signif==T][order(modmed_acme_p)] ) # None signif :(
knitr::kable(mwas[intrx_p<0.1, .(HMDB_ID,Metabolite,unique_met_id,modmed_acme_p,modmed_ade_p)][order(modmed_acme_p)] )
```
